---
title: "Microbiome data analysis with microViz"
author: "David Barnett"
format: 
  revealjs:
    theme: dark
    transition: fade
    slide-number: true
    scrollable: true
    pause: false
    fig-responsive: false
    history: false
    multiplex: true
editor: source
---

# Part 1: Diversity {transition="convex"}

```{r message=FALSE}
library(dplyr)
library(purrr)
library(ggplot2)
library(cowplot)
library(phyloseq)
library(microViz)
```

```{r}
knitr::opts_chunk$set(fig.width = 6, fig.height = 4, dpi = 120)
options(width = 500)
miceBars <- readRDS(here::here('slides/data/barplots.rds'))
```

::: notes
Okay.
So before the break we processed some amplicon sequencing data with DADA2.
We identified the unique bacterial taxa (ASVs) in a few samples, assigned the ASVs' taxonomic classifications by mapping to a reference database, and quantified their relative abundances.
We finished up by placing this data into a phyloseq object.

Now we're going to look at a phyloseq object containing some similar 16S amplicon sequencing data, which has been processed already with DADA2.
This data comes from an experimental study of the mouse gut microbiome.
There are more details on the study design and aims in the exercise material, but the details aren't important for now.
:::

## Intro to phyloseq {transition="convex"}

<br>

phyloseq S4 object: processed mouse gut microbiota data

```{r}
#| echo: true
mice <- readRDS(here::here('data/mice.rds'))
```

<br>

::: fragment
Printed object shows you functions to access the data inside!

```{r}
#| echo: true
mice
```
:::

::: notes
We read in the R data object, and print it.
Now lets look at the pieces.
:::

## Taxonomy table

```{r}
#| echo: true
#| output-location: fragment
tax_table(mice) %>% head(15)
```

## OTU table {.smaller}

```{r}
#| echo: true
otu_table(mice)[1:8, 1:8]
```

::: fragment
You can also use the \@ symbol.

```{r}
#| echo: true
mice@otu_table[1:8, 1:8] # the same result
```
:::

## Sample data (a.k.a. metadata) {.smaller}

<br>

```{r}
#| echo: true
sample_data(mice)[1:15, 1:10]
```

## 

Ranks

```{r}
#| echo: true
rank_names(mice)
```

Taxa

```{r}
#| echo: true
taxa_names(mice) %>% head() 
```

Samples

```{r}
#| echo: true
sample_names(mice) %>% head() 
```

Variables

```{r}
#| echo: true
sample_variables(mice)
```

## Cheat code {transition="convex-out"}

``` r
View(mice)
```

::: notes
Interactively in Rstudio, View is an easy way to look at S4 objects.
But don't put this in a script!
:::

# Seeing is believing {transition="convex-in convex-out"}

Get a feel for your data, by making simple plots.

We're going to use the package `microViz`

::: notes
Okay, so how do we look at the microbiota abundance data?
:::

## Data subsetting {transition="convex"}

We can subset the samples using the sample_data info

```{r}
#| echo: true
miceSubset <- mice %>% ps_filter(
  treatment_days == "D13", virus == "WNV2000", treatment == "Vehicle"
)
miceSubset
```

::: notes
Lets take a very small subset of this data to get started, just the control group at day 13.
:::

## Basic bar plots {.smaller transition="fade"}

``` r
miceSubset %>% comp_barplot(
  tax_level = "unique", n_taxa = 12, bar_width = 0.7,
  sample_order = "asis", tax_transform_for_plot = "identity"
) + coord_flip()
```

```{r}
miceBars$bad %>% ggdraw()
```

::: notes
We'll use the function comp_barplot to make bar plots of the microbiome composition.

Each row of the plot is a different sample.
And the bar heights represent the readcounts from each sample which are coloured by the ASV they belong to.
But we can improve this plot.
The ASVs have uninformative IDs, and the total number of reads is not particularly useful information.
:::

## Proportional bar plots {.smaller transition="fade"}

``` r
miceSubset %>% comp_barplot(
  tax_level = "unique", n_taxa = 12, bar_width = 0.7,
  sample_order = "asis"
) + coord_flip()
```

```{r}
miceBars$comp %>% ggdraw()
```

::: notes
Total readcount from the sequencing does not correspond to the total biomass or bacterial load of the samples.
So for now we will just consider the relative abundance of each taxon, as proportions of the total counts for that sample.
:::

```{r}
mice <- tax_fix(mice, verbose = FALSE)
mice <- tax_rename(mice, rank = 'Family')
miceSubset <- mice %>% 
  ps_filter(treatment_days == "D13", virus == "WNV2000", treatment == "Vehicle")
```

## Proportional bar plot - better names {.smaller transition="fade"}

``` r
miceSubset %>% comp_barplot(
  tax_level = 'unique', n_taxa = 12, bar_width = 0.7,
  sample_order = 'asis'
) + coord_flip()
```

```{r}
miceBars$named %>% ggdraw()
```

::: notes
We can also rename the unique taxa with a more informative name, according to their classification at, for example, Family level, and how common they are.
Helpful, now we can see that most of the top ASVs are from the Porphyromonadaceae family.
:::

## Proportional bar plot - all ASVs {.smaller transition="fade"}

``` r
miceSubset %>% comp_barplot(
  tax_level = 'unique', n_taxa = 12, bar_width = 0.7,
  sample_order = 'asis', merge_other = FALSE
) + coord_flip()
```

```{r}
miceBars$unique %>% ggdraw()
```

::: notes
But sadly we don't have enough distinct colours to label all the unique taxa, we can show them like this.
There's a lot.
:::

## Aggregated bar plot - Families {.smaller transition="fade"}

``` r
miceSubset %>% comp_barplot(
  tax_level = "Family", n_taxa = 10, bar_width = 0.7, 
  sample_order = 'asis'
) + coord_flip()
```

```{r}
miceBars$family %>% ggdraw()
```

::: notes
So let's "aggregate" all the counts into family-level groups.
By aggregating at family level, we have sacrificed taxonomic resolution, compared to using ASVs.
But this way we can get an idea of which families are the most abundant, and how variable the communities are.
You can make similar plots aggregated at different taxonomic ranks.
:::

## Aggregated bar plot - Genera {.smaller transition="fade"}

``` r
miceSubset %>% comp_barplot(
    tax_level = "Genus", n_taxa = 12, bar_width = 0.7,
    sample_order = 'asis', merge_other = FALSE
  ) + coord_flip()
```

```{r}
miceBars$genus %>% ggdraw()
```

::: notes
Many of the ASVs in this mice data, the *Porphyromonadaceae*, could not be classified at genus level.
:::

## Aggregated bar plot - Phyla {.smaller transition="fade-in convex-out"}

``` r
miceSubset %>% comp_barplot(
    tax_level = "Phylum", n_taxa = 7, bar_width = 0.7, 
    sample_order = 'asis'
  ) + coord_flip()
```

```{r}
miceBars$phylum %>% ggdraw()
```

::: notes
A note on phylum names!
Next slide.
:::

## Phylum name changes! {transition="convex"}

::: incremental
-   *Actinobacteria* is now *Actinomycetota*
-   *Bacteroidetes* is now *Bacteroidota*
-   *Proteobacteria* is now *Pseudomonadota* (!)
-   *Firmicutes* is now *Bacillota* (!?!)
:::

::: notes
There have been major changes this year and some of these are now old names.
Most published research is of course with the old names (and still probably will be for a year or so).
:::

# Alpha diversity {transition="convex-in convex-out"}

How diverse is the bacterial microbiome of each sample?

## Why is diversity interesting?

<br>

#### Biologically

-   Diverse == healthy, resilient, good (?)

<br>

#### Practically

-   Summarize an ecosystem in one number: easy to compare

::: notes
-   Lower gut microbiome diversity is related to worse health in adult humans.
-   Higher diversity ecosystems are often considered healthier, more mature, and more resilient to perturbation.
-   BUT: diverse == healthy does not hold for all ecosystems, e.g. early infant gut microbiome, so consider your own data and hypotheses carefully.
:::

# Quantifying diversity? {transition="convex"}

## Observed Richness

```{r}
#| out-width: '6.6in'
#| out-height: '4.4in'
mice %>%
  ps_calc_richness(rank = "Family", index = "observed", varname = "N") %>%
  ps_dedupe(vars = "N", method = "first", verbose = FALSE) %>%
  ps_filter(N %% 2 == 1) %>%
  ps_arrange(desc(N)) %>%
  comp_barplot(
    tax_level = "Family", n_taxa = 13, label = "N", bar_width = 0.7,
    sample_order = "asis", merge_other = FALSE
  ) + coord_flip()
```

The more the merrier.
Just counting.
$N = N$.

::: notes
The simplest measure of diversity, or more strictly "richness" is just counting, aka "Observed Richness".
Here we've computed the observed richness of family level taxa and labeled each sample with it.
I took a range of samples with different richnesses, so you can get an idea how this looks.
:::

## Diversity - Shannon index (H)

```{r}
#| out-height: '4.4in'
#| out-width: '6.6in'
mice %>%
  ps_calc_richness(rank = "Family", index = "observed", varname = "N") %>%
  ps_calc_diversity(rank = "Family", index = "shannon", varname = "shannon") %>%
  ps_dedupe(vars = "N", method = "first", verbose = FALSE) %>%
  ps_filter(N %% 2 == 1) %>%
  ps_arrange(desc(shannon)) %>%
  ps_mutate(shannon = sprintf("%.1f", shannon)) %>%
  comp_barplot(
    tax_level = "Family", n_taxa = 13, label = "shannon", bar_width = 0.7,
    sample_order = "asis", merge_other = FALSE
  ) + coord_flip()
```

Richness AND evenness matter.
$H = -\sum_{i=1}^Np_i\ln p_i$

::: notes
A better measure of ecosystem diversity will also consider the evenness of the ecosystem.
If an ecosystem is dominated by one taxon, with several other taxa present, but which only a few counts, that is intuitively a less diverse ecosystem than one with an even distribution of the same number of taxa.
You can kind of see this from the bars which are the same samples as before, just now ordered and labelled by increasing Shannon diveristy.

Shannon index often is called H, p is proportion of the i'th taxon.
For each taxon, you multiply the proportional abundance by the natural log of the proportion.
You can try it out for yourself to convince yourself you get larger (negative) values for higher proportions.
The highest value you can acheive with for example 20 taxa, is if they each have a proportion of 5%.
You change the sign of the result to a positive number, as this makes more sense: higher numbers means higher diversity.
:::

## Diversity - Effective numbers

```{r}
#| out-height: '4.4in'
#| out-width: '6.6in'
mice %>%
  ps_calc_richness(rank = "Family", index = "observed", varname = "N") %>%
  ps_calc_diversity(rank = "Family", index = "shannon", varname = "shannon") %>%
  ps_dedupe(vars = "N", method = "first", verbose = FALSE) %>%
  ps_filter(N %% 2 == 1) %>%
  ps_arrange(desc(shannon)) %>%
  ps_mutate(shannon = round(exp(shannon))) %>%
  comp_barplot(
    tax_level = "Family", n_taxa = 13, label = "shannon", bar_width = 0.7,
    sample_order = "asis", merge_other = FALSE
  ) + coord_flip()
```

$e^H = N$ if all taxa were equally abundant.

::: notes
The value of the Shannon index itself has no intuitive meaning.
You can compare them, but not so easily interpret any one number.

The concept of "effective numbers" of taxa is useful here.
If you take the exponent of the Shannon index, what you get is the number of taxa in an evenly abundant ecosystem with the same Shannon index.
This will be the actual number of taxa (observed richness) if your original ecosystem was actually perfectly even.
It will otherwise be less than the number of taxa actually in your ecosystem.
:::

## Your turn {transition="convex"}

-   Link to exercises guidance: <https://david-barnett.github.io/evomics-material/exercises/microViz-1-intro-diversity-exercises>

-   Try out exploring the mouse data, making bar charts, calculating alpha diversity and doing some simple stats

-   Excellent resource for more diversity details: <https://www.davidzeleny.net/anadat-r/doku.php/en:div-ind>

-   Next topic after dinner: Dissimilarity and Ordination
