<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.313">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="David Barnett">
<meta name="dcterms.date" content="2023-05-24">

<title>Evomics Workshop part 1</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="exercises_1_files/libs/clipboard/clipboard.min.js"></script>
<script src="exercises_1_files/libs/quarto-html/quarto.js"></script>
<script src="exercises_1_files/libs/quarto-html/popper.min.js"></script>
<script src="exercises_1_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="exercises_1_files/libs/quarto-html/anchor.min.js"></script>
<link href="exercises_1_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="exercises_1_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="exercises_1_files/libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="exercises_1_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="exercises_1_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="exercises_1_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="exercises_1_files/libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#instructions" id="toc-instructions" class="nav-link active" data-scroll-target="#instructions">Instructions üëã</a></li>
  <li><a href="#setup" id="toc-setup" class="nav-link" data-scroll-target="#setup">Setup ‚öôÔ∏è</a></li>
  <li><a href="#intro-to-phyloseq" id="toc-intro-to-phyloseq" class="nav-link" data-scroll-target="#intro-to-phyloseq">Intro to phyloseq ü¶†</a>
  <ul>
  <li><a href="#filling-gaps-in-the-taxonomy-table" id="toc-filling-gaps-in-the-taxonomy-table" class="nav-link" data-scroll-target="#filling-gaps-in-the-taxonomy-table">Filling gaps in the taxonomy table</a></li>
  </ul></li>
  <li><a href="#plot-some-microbiome-compositions" id="toc-plot-some-microbiome-compositions" class="nav-link" data-scroll-target="#plot-some-microbiome-compositions">Plot some microbiome compositions!</a>
  <ul>
  <li><a href="#stacked-compositional-barplots" id="toc-stacked-compositional-barplots" class="nav-link" data-scroll-target="#stacked-compositional-barplots">Stacked compositional barplots</a></li>
  <li><a href="#aggregating-taxa" id="toc-aggregating-taxa" class="nav-link" data-scroll-target="#aggregating-taxa">Aggregating taxa</a></li>
  <li><a href="#organising-your-bar-charts" id="toc-organising-your-bar-charts" class="nav-link" data-scroll-target="#organising-your-bar-charts">Organising your bar charts</a></li>
  </ul></li>
  <li><a href="#ecosystem-diversity" id="toc-ecosystem-diversity" class="nav-link" data-scroll-target="#ecosystem-diversity">Ecosystem diversity</a>
  <ul>
  <li><a href="#richness" id="toc-richness" class="nav-link" data-scroll-target="#richness">Richness</a></li>
  <li><a href="#diversity" id="toc-diversity" class="nav-link" data-scroll-target="#diversity">Diversity</a></li>
  </ul></li>
  <li><a href="#comparing-diversity" id="toc-comparing-diversity" class="nav-link" data-scroll-target="#comparing-diversity">Comparing Diversity</a></li>
  <li><a href="#break-time" id="toc-break-time" class="nav-link" data-scroll-target="#break-time">Break Time! ‚òïÔ∏è ‚åöÔ∏è</a>
  <ul>
  <li><a href="#you-want-more-about-diversity" id="toc-you-want-more-about-diversity" class="nav-link" data-scroll-target="#you-want-more-about-diversity">You want more about diversity?</a></li>
  </ul></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session info</a></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Evomics Workshop part 1</h1>
<p class="subtitle lead">Microbiome analysis with microViz</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>David Barnett </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">May 24, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="instructions" class="level2">
<h2 class="anchored" data-anchor-id="instructions">Instructions üëã</h2>
<ul>
<li>These exercises will introduce some core concepts in microbiome analysis, using example data</li>
<li>Follow along by writing code in your own R script (or an R Markdown notebook, or a Quarto doc)</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Code blocks have a copy button in the top right -&gt; -&gt; -&gt; -&gt; -&gt;</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"- Copy and modify the code examples</span><span class="sc">\n</span><span class="st">- Try to understand what it does, and play around!"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>- Copy and modify the code examples
- Try to understand what it does, and play around!</code></pre>
</div>
</div>
<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Click on these collapsed callout blocks for more info
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><strong>Topics for part 1:</strong></p>
<ol type="1">
<li>phyloseq objects for microbiome data</li>
<li>Stacked bar charts for visualizing microbiome compositions</li>
<li>Ecosystem diversity indices</li>
</ol>
</div>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Getting help
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The <a href="https://david-barnett.github.io/microViz/">microViz website</a> has help pages for every function and more tutorials.</p>
</div>
</div>
</div>
<hr>
</section>
<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup ‚öôÔ∏è</h2>
<ol type="1">
<li>Open your RStudio server instance (or local RStudio if you prefer)</li>
<li>Find the microbiome analysis folder</li>
<li>Click on the .Rproj file to open an RStudio project (recommended)</li>
<li>Open a new R script / notebook / quarto doc</li>
<li>Load the R packages we will be using:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microViz)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(shiny)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="6" type="1">
<li>Explore some microbiomes</li>
</ol>
<ul>
<li>We‚Äôll use two example datasets in this evening‚Äôs exercises</li>
<li>Both contain gut microbiome taxonomic composition data, from stool samples</li>
<li>The first is from a 2019 study of antibiotic administration in mice</li>
<li>The mouse antibiotics study used 16S rRNA gene amplicon sequencing
<ul>
<li>They used Illumina MiSeq and processed the data into ASVs using DADA2</li>
</ul></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check data_directory is correct path on AMI to data folder</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>data_directory <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>) </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>mice <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(data_directory, <span class="st">"mice.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>The second is from a large birth cohort study, containing human infants born by C-section or vaginal delivery</li>
<li>The birth cohort study data is shotgun metagenomics data
<ul>
<li>They used Illumina HiSeq and inferred the abundance of species-like features with metaphlan3.</li>
</ul></li>
</ul>
<!-- (Note: not metaphlan3 in the paper, but available from curatedMetagenomicData package/project) -->
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"shao19"</span>, <span class="at">package =</span> <span class="st">"microViz"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>All analysis approaches we‚Äôll learn this evening can be applied to both kinds of data.</li>
<li>These methods are also appropriate for microbiome data from other environments, not just the gut microbiota! (microViz has been used by groups analysing microbiomes associated with Antarctic sponges, and Kiwi birds, amongst other things.)</li>
</ul>
</section>
<section id="intro-to-phyloseq" class="level2">
<h2 class="anchored" data-anchor-id="intro-to-phyloseq">Intro to phyloseq ü¶†</h2>
<p>We will start with some processed microbiota data from the mouse study.</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Click here for details about the mice dataset:
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>The data originate from <a href="https://www.ncbi.nlm.nih.gov/pubmed/29590614">a study on the effects of oral antibiotics on West Nile Virus (WNV) infection</a></li>
<li>Sequence data was generated from extracted DNA from mouse stool samples.</li>
<li>DNA was amplified using primers specific for the 16S rRNA gene V4 region using primers 515F/806R.</li>
<li>Each mouse was caged separately.</li>
</ul>
<div style="display: flex;">
<div style="width: 70%;">
<p><strong>The study followed WNV infection after the following treatments:</strong></p>
<ol type="1">
<li><strong>Ampicillin</strong> (Amp): <a href="https://en.wikipedia.org/wiki/Ampicillin" class="uri">https://en.wikipedia.org/wiki/Ampicillin</a></li>
<li><strong>Metronidazole</strong> (Metro): <a href="https://en.wikipedia.org/wiki/Metronidazole" class="uri">https://en.wikipedia.org/wiki/Metronidazole</a></li>
<li><strong>Ampicillin</strong> + <strong>Metronidazole</strong> (AmpMetro)</li>
<li><strong>Koolaid Control</strong> (Vehicle): <em>Antibiotics taste bad, so Koolaid is added as a sweetener!</em></li>
</ol>
</div>
<div style="width: 30%;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/lab-mouse.jpg" class="img-fluid figure-img" style="width:1.2in" alt="Lab mouse enjoying Koolaid. Imagined by Lexica."></p>
<p></p><figcaption class="figure-caption"><em>Yum, Koolaid!</em></figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>Treatments were supplied ad libitum for 2 weeks prior to viral infection and maintained for 2 weeks post-infection.</p>
<ul>
<li>The primary outcome of the study was mouse survival.</li>
<li>Each treatment group had two subgroups:
<ol type="a">
<li>mice infected with WNV, via a subcutaneous foot pad injection</li>
<li>mice left uninfected, as another set of controls (we will not use these samples at the start)</li>
</ol></li>
</ul>
</div>
</div>
</div>
<p><code>mice</code> is a <code>phyloseq</code> S4 object, containing a small subset of these data.</p>
<ul>
<li>Printing the object shows you some basic info.</li>
<li>It also names three functions you can use to access the data inside.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>mice</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 591 taxa and 45 samples ]
sample_data() Sample Data:       [ 45 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 591 taxa by 6 taxonomic ranks ]</code></pre>
</div>
</div>
<ul>
<li>Have a look around inside the different parts of this object using accessor functions.</li>
<li>phyloseq and microViz both provide accessor functions, some to try are included below.</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">samdat_tbl</span>()</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_variables</span>()</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">otu_get</span>()</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">nsamples</span>()</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ntaxa</span>()</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_names</span>()</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="fu">taxa_names</span>() </span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># don't forget you can use head() to truncate long output!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Tip: get an interactive View()
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>You can also try using the <code>View()</code> function on each part</li>
<li>But remember not to leave interactive functions like <code>View</code> in your script!</li>
</ul>
</div>
</div>
<div class="callout-warning callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
The @ symbol, a risky shortcut?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>You can also use the <code>@</code> symbol to access the slots of S4 objects e.g.&nbsp;phyloseq</li>
<li>This can be a quick shortcut when exploring a data object interactively</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>mice<span class="sc">@</span>otu_table[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>OTU Table:          [5 taxa and 5 samples]
                     taxa are columns
      ASV0001 ASV0002 ASV0003 ASV0004 ASV0005
D7.B1     709      33    1715     224     555
D7.B2    3896      22    7808    1201    4306
D7.B3     494      13    2657    4841     175
D7.B4    3147      80    4359    6269    3795
D7.B5    1663      11    3577     283    1814</code></pre>
</div>
</div>
<div class="sourceCode" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>mice<span class="sc">@</span>sam_data <span class="sc">%&gt;%</span> <span class="fu">View</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>But, it is slightly risky to directly access the internal structure of S4 objects!</li>
<li>Package authors may change the object‚Äôs internal structure when they release a new version, breaking your code.</li>
<li><strong>So in your scripts, prefer accessor functions over <code>@</code></strong></li>
</ul>
</div>
</div>
</div>
<section id="filling-gaps-in-the-taxonomy-table" class="level3">
<h3 class="anchored" data-anchor-id="filling-gaps-in-the-taxonomy-table">Filling gaps in the taxonomy table</h3>
<ul>
<li>You might have noticed the the taxonomy table has some <code>NA</code> values.</li>
<li>This often occurs when an ASV cannot be uniquely classified at the Genus level.</li>
<li>The short 16S region sequenced may only allow a unique classification at Family rank or above.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tax_table</span>(mice)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Taxonomy Table:     [5 taxa by 6 taxonomic ranks]:
        Kingdom    Phylum           Class                 Order            
ASV0001 "Bacteria" "Bacteroidetes"  "Bacteroidia"         "Bacteroidales"  
ASV0002 "Bacteria" "Proteobacteria" "Gammaproteobacteria" "Xanthomonadales"
ASV0003 "Bacteria" "Bacteroidetes"  "Bacteroidia"         "Bacteroidales"  
ASV0004 "Bacteria" "Bacteroidetes"  "Bacteroidia"         "Bacteroidales"  
ASV0005 "Bacteria" "Bacteroidetes"  "Bacteroidia"         "Bacteroidales"  
        Family               Genus             
ASV0001 "Porphyromonadaceae" NA                
ASV0002 "Xanthomonadaceae"   "Stenotrophomonas"
ASV0003 "Porphyromonadaceae" NA                
ASV0004 "Porphyromonadaceae" NA                
ASV0005 "Porphyromonadaceae" NA                </code></pre>
</div>
</div>
<ul>
<li><p>We need to fill those gaps! and we can do this with <code>tax_fix()</code></p></li>
<li><p><code>tax_fix()</code> copies info down from a higher rank, to fill the gaps.</p></li>
<li><p>First, let‚Äôs look at the taxonomy table interactively, with <code>tax_fix_interactive()</code></p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tax_fix_interactive(mice) # run this in the R Console </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout-warning callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Press Stop!
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Running <code>tax_fix_interactive(mice)</code> will open a new web browser tab.</li>
<li>When you are done looking, click the <strong>red STOP button in the R console</strong>!</li>
</ul>
</div>
</div>
<ul>
<li>Let‚Äôs update our <code>mice</code> phyloseq object with this fix.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>mice <span class="ot">&lt;-</span> <span class="fu">tax_fix</span>(mice, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Check the first few taxa now look correct?</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Hint: use tax_table and head</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr>
</section>
</section>
<section id="plot-some-microbiome-compositions" class="level2">
<h2 class="anchored" data-anchor-id="plot-some-microbiome-compositions">Plot some microbiome compositions!</h2>
<ul>
<li>Okay, so how do we look at the microbiota abundance data?</li>
<li>We‚Äôre first going to make stacked compositional bar charts using the <code>microViz</code> package, <code>comp_barplot()</code> function.</li>
</ul>
<section id="stacked-compositional-barplots" class="level3">
<h3 class="anchored" data-anchor-id="stacked-compositional-barplots">Stacked compositional barplots</h3>
<ul>
<li>Lets start with a smaller subset of the data.</li>
<li>Just the control group (Vehicle treatment) at day 13.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can filter the samples like this, using the sample_data information</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>mice <span class="sc">%&gt;%</span> <span class="fu">ps_filter</span>(treatment <span class="sc">==</span> <span class="st">"Vehicle"</span>) <span class="co"># similar to a dplyr filter!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 394 taxa and 10 samples ]
sample_data() Sample Data:       [ 10 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 394 taxa by 6 taxonomic ranks ]</code></pre>
</div>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Tip: Forgot the variable‚Äôs levels?
</div>
</div>
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use table()</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>mice<span class="sc">@</span>sam_data<span class="sc">$</span>treatment <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>mice <span class="sc">%&gt;%</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_filter</span>(treatment <span class="sc">==</span> <span class="st">"Vehicle"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_names2rank</span>(<span class="st">"ASV"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_barplot</span>(</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_level =</span> <span class="st">"ASV"</span>, <span class="at">n_taxa =</span> <span class="dv">12</span>, </span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">bar_width =</span> <span class="fl">0.7</span>, <span class="at">sample_order =</span> <span class="st">"asis"</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_1_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<ul>
<li>The ASVs have uninformative numeric IDs, but we can fix that with <code>tax_rename()</code></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>mice <span class="sc">%&gt;%</span> </span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_filter</span>(treatment <span class="sc">==</span> <span class="st">"Vehicle"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_rename</span>(<span class="at">rank =</span> <span class="st">"Family"</span>, <span class="at">pad_digits =</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span> <span class="co"># this changes the taxa_names</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_names2rank</span>(<span class="st">"ASV"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_barplot</span>(</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_level =</span> <span class="st">"ASV"</span>, <span class="at">n_taxa =</span> <span class="dv">12</span>, </span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">bar_width =</span> <span class="fl">0.7</span>, <span class="at">sample_order =</span> <span class="st">"asis"</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_1_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Tip: What is that function?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>When you see a function you are unfamiliar with, e.g.&nbsp;<code>tax_names2rank()</code>, look for documentation:</p>
<ol type="1">
<li>Run <code>?tax_names2rank</code> in the console, or click on the function and press the F1 key (F2 shows the source code!)</li>
<li>Many packages have nice websites: search for the function reference page, e.g.&nbsp;<a href="https://david-barnett.github.io/microViz/reference/index.html" class="uri">https://david-barnett.github.io/microViz/reference/index.html</a></li>
<li>Test out the function in isolation, does it do what you expected?</li>
</ol>
<div class="sourceCode" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>mice <span class="sc">%&gt;%</span> </span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_rename</span>(<span class="at">rank =</span> <span class="st">"Family"</span>, <span class="at">pad_digits =</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">taxa_names</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">12</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
<section id="aggregating-taxa" class="level3">
<h3 class="anchored" data-anchor-id="aggregating-taxa">Aggregating taxa</h3>
<ul>
<li>Sadly we don‚Äôt have enough distinct colours to show all the unique ASVs.</li>
<li>Instead we can <strong>aggregate</strong> the counts at a higher taxonomic rank, e.g.&nbsp;Families.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>mice <span class="sc">%&gt;%</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_filter</span>(treatment <span class="sc">==</span> <span class="st">"Vehicle"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_barplot</span>(</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_level =</span> <span class="st">"Family"</span>, <span class="at">n_taxa =</span> <span class="dv">9</span>, </span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample_order =</span> <span class="st">"asis"</span>, <span class="at">bar_width =</span> <span class="fl">0.7</span>,</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">merge_other =</span> <span class="cn">FALSE</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_1_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Tip: Use ggsave to save your plots?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>Viewing your plots in the RStudio Plots window is okay for practice</li>
<li>But don‚Äôt copy paste them to save them, as there is a much better way!</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign your plot to an R object</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>miceBarsD7koolaid <span class="ot">&lt;-</span> mice <span class="sc">%&gt;%</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_filter</span>(treatment <span class="sc">==</span> <span class="st">"Vehicle"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_barplot</span>(</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_level =</span> <span class="st">"Family"</span>, <span class="at">n_taxa =</span> <span class="dv">9</span>, </span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample_order =</span> <span class="st">"asis"</span>, <span class="at">bar_width =</span> <span class="fl">0.7</span>,</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">merge_other =</span> <span class="cn">FALSE</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  ) </span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Write the plot to a file, with ggsave</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">filename =</span> <span class="st">"mice-barchart-day7-WNV.png"</span>, <span class="at">device =</span> <span class="st">"png"</span>,</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> miceBarsD7koolaid, <span class="at">width =</span> <span class="fl">6.5</span>, <span class="at">height =</span> <span class="fl">3.5</span>, <span class="at">dpi =</span> <span class="dv">200</span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Be sure to carefully adjust the sizing and resolution of your plots for your paper or presentation!</strong></p>
</div>
</div>
</div>
<ul>
<li>Try aggregating at other ranks, e.g.&nbsp;Class.</li>
</ul>
<div class="callout-caution callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Fickle Phyla, beware!
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>There were major changes to bacterial taxonomy last year</li>
<li>Many of the phylum names in our dataset are now officially outdated.</li>
<li>Currently, some research is published with the old names and some with the new names.</li>
<li>Here are the key changes to be aware of for the human gut microbiota
<ul>
<li><em>Actinobacteria</em> is now <em>Actinomycetota</em></li>
<li><em>Bacteroidetes</em> is now <em>Bacteroidota</em></li>
<li><em>Proteobacteria</em> is now <em>Pseudomonadota</em> (!)</li>
<li><em>Firmicutes</em> is now <em>Bacillota</em> (!?!)</li>
</ul></li>
</ul>
</div>
</div>
</div>
</section>
<section id="organising-your-bar-charts" class="level3">
<h3 class="anchored" data-anchor-id="organising-your-bar-charts">Organising your bar charts</h3>
<ul>
<li>Let‚Äôs look at all of the data from day 7, from all treatment groups.</li>
<li>We can add the ggplot2 <code>facet_wrap()</code> to our plot, to separate the treatment groups.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>mice <span class="sc">%&gt;%</span> </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_barplot</span>(</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_level =</span> <span class="st">"Family"</span>, <span class="at">n_taxa =</span> <span class="dv">10</span>, <span class="at">merge_other =</span> <span class="cn">FALSE</span>,</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample_order =</span> <span class="st">"asis"</span>, <span class="at">label =</span> <span class="st">"cage"</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="at">facets =</span> <span class="fu">vars</span>(treatment), <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Day 7 mouse fecal microbiota compositions"</span>,</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Mouse ID"</span>, <span class="at">y =</span> <span class="st">"Relative Abundance"</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_1_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="576"></p>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># coord_flip exchanges x and y axes, which can be more readable</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Optional extension exercise: More mice data!
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><strong>Not recommended for R or ggplot2 beginners.</strong></p>
<p>This optional plotting exercise is more complex, and aims to exemplify the kind of complex manual arrangements enabled by comp_barplot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Additional packages needed</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>So far we looked just at one day of measurements, but stool samples were collected on multiple days.</li>
<li>Let‚Äôs grab the full dataset for more practice arranging our bar charts.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>allMice <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(data_directory, <span class="st">"allMice.rds"</span>))</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>allMice <span class="ot">&lt;-</span> <span class="fu">tax_fix</span>(allMice, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>allMice</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 3229 taxa and 520 samples ]
sample_data() Sample Data:       [ 520 samples by 11 sample variables ]
tax_table()   Taxonomy Table:    [ 3229 taxa by 6 taxonomic ranks ]</code></pre>
</div>
</div>
<div class="cell">
<details>
<summary>Code to set up a grid layout using cage numbers and letters</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert sample timings to a factor variable, for correct temporal ordering. </span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>allMice <span class="ot">&lt;-</span> allMice <span class="sc">%&gt;%</span> </span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_mutate</span>(<span class="at">Days =</span> <span class="fu">factor</span>(</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> treatment_days,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">levels =</span> <span class="fu">paste0</span>(<span class="st">"D"</span>, <span class="fu">c</span>(<span class="st">".14"</span>, <span class="st">"0"</span>, <span class="st">"3"</span>, <span class="st">"7"</span>, <span class="st">"13"</span>, <span class="st">"16"</span>, <span class="st">"18"</span>, <span class="st">"20"</span>)),</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"-14"</span>, <span class="st">"0"</span>, <span class="st">"3"</span>, <span class="st">"7"</span>, <span class="st">"13"</span>, <span class="st">"16"</span>, <span class="st">"18"</span>, <span class="st">"20"</span>)</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate the cage numbers and letters to allow a grid layout.</span></span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>allMice <span class="ot">&lt;-</span> allMice <span class="sc">%&gt;%</span> </span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_mutate</span>(</span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">cage_number =</span> stringr<span class="sc">::</span><span class="fu">str_extract</span>(cage, <span class="st">"[1-9]"</span>),</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">cage_letter =</span> stringr<span class="sc">::</span><span class="fu">str_extract</span>(cage, <span class="st">"[A-Z]"</span>)</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the grid layout: plot the virus infected, vehicle control group</span></span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>allMice <span class="sc">%&gt;%</span> </span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_filter</span>(virus <span class="sc">==</span> <span class="st">"WNV2000"</span>, treatment <span class="sc">==</span> <span class="st">"Vehicle"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_barplot</span>(</span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_level =</span> <span class="st">"Family"</span>, <span class="at">n_taxa =</span> <span class="dv">13</span>, <span class="at">x =</span> <span class="st">'Days'</span>,</span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">bar_width =</span> <span class="fl">0.8</span>, <span class="at">merge_other =</span> <span class="cn">FALSE</span></span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="at">cols =</span> <span class="fu">vars</span>(cage_number), <span class="at">rows =</span> <span class="fu">vars</span>(cage_letter))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Registered S3 method overwritten by 'seriation':
  method         from 
  reorder.hclust vegan</code></pre>
</div>
<div class="cell-output-display">
<p><img src="exercises_1_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<details>
<summary>Code to create a big plot of all samples from WNV-exposed mice</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a list of identically themed and coloured plots with the group_by argument.</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>plotList <span class="ot">&lt;-</span> allMice <span class="sc">%&gt;%</span> </span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_mutate</span>(<span class="at">group =</span> <span class="fu">paste</span>(treatment, virus)) <span class="sc">%&gt;%</span> </span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_barplot</span>(<span class="at">tax_level =</span> <span class="st">"Family"</span>, <span class="at">n_taxa =</span> <span class="dv">13</span>, <span class="at">x =</span> <span class="st">"Days"</span>, <span class="at">group_by =</span> <span class="st">"group"</span>)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Arrange the WNV-exposed mice plots with patchwork, in 4 rows</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>plotsWNV <span class="ot">&lt;-</span> <span class="fu">wrap_plots</span>(plotList[<span class="fu">grep</span>(<span class="at">x =</span> <span class="fu">names</span>(plotList), <span class="at">pattern =</span> <span class="st">"WNV2000"</span>)]) </span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>plotsWNV <span class="ot">&lt;-</span> plotsWNV <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>, <span class="at">nrow =</span> <span class="dv">4</span>, <span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">2</span>))</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Add faceting to all the plots in the list, with patchwork's `&amp;` operator</span></span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>plotsWNV <span class="ot">&lt;-</span> plotsWNV <span class="sc">&amp;</span> <span class="fu">facet_grid</span>(<span class="at">cols =</span> <span class="fu">vars</span>(cage_number), <span class="at">rows =</span> <span class="fu">vars</span>(cage_letter))</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Write these plots to a large image file!</span></span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">plot =</span> plotsWNV, <span class="at">filename =</span> <span class="st">"WNVplots.png"</span>, <span class="at">height =</span> <span class="dv">12</span>, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">dpi =</span> <span class="dv">200</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details>
<summary>Exercise - write code for a plot of all control mice (no virus)</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</div>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-12-contents" aria-controls="callout-12" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
More barchart resources:
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-12" class="callout-12-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>More examples/tutorial of visualizing microbiome data using stacked barcharts can be found here: <a href="https://david-barnett.github.io/microViz/articles/web-only/compositions.html" class="uri">https://david-barnett.github.io/microViz/articles/web-only/compositions.html</a></p>
<p><strong>Note: Bar charts often look better when you sort the samples by similarity.</strong></p>
<ul>
<li>The webpage mentions using Bray-Curtis distances and hierarchical clustering to sort samples.</li>
<li>We haven‚Äôt discussed dissimilarity or distances yet, but we will in the next set of exercises!</li>
<li>For now, just appreciate that it can make it easier to see patterns in your compositions!</li>
</ul>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="ecosystem-diversity" class="level2">
<h2 class="anchored" data-anchor-id="ecosystem-diversity">Ecosystem diversity</h2>
<p>How diverse is the bacterial microbiome of each sample?</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Why is diversity interesting?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="biologically" class="level4">
<h4 class="anchored" data-anchor-id="biologically">Biologically</h4>
<ul>
<li>Higher diversity ecosystems are probably more resilient to perturbations</li>
<li>Lower gut microbiome diversity is related to worse health in adult humans</li>
<li>BUT: <code>diverse == healthy</code> is not <code>TRUE</code> for all ecosystems (e.g.&nbsp;early infant gut microbiome)</li>
<li>So, consider your own data and diversity hypotheses carefully</li>
</ul>
</section>
<section id="practically" class="level4">
<h4 class="anchored" data-anchor-id="practically">Practically</h4>
<ul>
<li>Diversity indices provide a simple one number summary of each ecosystem</li>
<li>This makes it relatively easy to compare samples, and do statistical testing</li>
</ul>
</section>
</div>
</div>
</div>
<section id="richness" class="level3">
<h3 class="anchored" data-anchor-id="richness">Richness</h3>
<ul>
<li>The simplest richness measure is just counting, a.k.a. ‚ÄúObserved Richness‚Äù.</li>
<li>Let‚Äôs compute the observed richness and sort and label some samples.</li>
<li><code>ps_calc_richness()</code> computes the index for each sample and adds it to your sample_data</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>mice <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_filter</span>(treatment <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Vehicle"</span>, <span class="st">"AmpMetro"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_calc_richness</span>(<span class="at">rank =</span> <span class="st">"Family"</span>, <span class="at">index =</span> <span class="st">"observed"</span>, <span class="at">varname =</span> <span class="st">"N_Families"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_arrange</span>(N_Families) <span class="sc">%&gt;%</span> </span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_barplot</span>(</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_level =</span> <span class="st">"Family"</span>, <span class="at">n_taxa =</span> <span class="dv">14</span>, <span class="at">label =</span> <span class="st">"N_Families"</span>, <span class="at">bar_width =</span> <span class="fl">0.8</span>,</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample_order =</span> <span class="st">"asis"</span>, <span class="at">merge_other =</span> <span class="cn">FALSE</span>, <span class="at">facet_by =</span> <span class="st">"treatment"</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_1_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="576"></p>
</div>
</div>
</section>
<section id="diversity" class="level3">
<h3 class="anchored" data-anchor-id="diversity">Diversity</h3>
<ul>
<li>A true measure of ecosystem diversity will also consider the evenness of the ecosystem.</li>
<li>A rich ecosystem predominated by one taxon is still intuitively a less diverse ecosystem than one with an even distribution of the same number of taxa.</li>
</ul>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true" href="">1. Shannon Index</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false" href="">2. Effective Shannon Diversity</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="cell">
<details>
<summary>Code to compute Shannon Index and make a barplot</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>mice <span class="sc">%&gt;%</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_filter</span>(treatment <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Vehicle"</span>, <span class="st">"AmpMetro"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_calc_diversity</span>(<span class="at">rank =</span> <span class="st">"Family"</span>, <span class="at">index =</span> <span class="st">"shannon"</span>, <span class="at">varname =</span> <span class="st">"Shannon_Family"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_arrange</span>(Shannon_Family) <span class="sc">%&gt;%</span> </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_mutate</span>(<span class="at">Shannon =</span> <span class="fu">formatC</span>(Shannon_Family, <span class="at">digits =</span> <span class="dv">2</span>, <span class="at">format =</span> <span class="st">"f"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_barplot</span>(</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_level =</span> <span class="st">"Family"</span>, <span class="at">n_taxa =</span> <span class="dv">14</span>, <span class="at">label =</span> <span class="st">"Shannon"</span>, <span class="at">bar_width =</span> <span class="fl">0.8</span>,</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample_order =</span> <span class="st">"asis"</span>, <span class="at">merge_other =</span> <span class="cn">FALSE</span>, <span class="at">facet_by =</span> <span class="st">"treatment"</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="exercises_1_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<ul>
<li>The Shannon index is a commonly used diversity measure, with this formula: <span class="math inline">\(H = -\sum_{i=1}^Np_i\ln p_i\)</span></li>
<li>Shannon index is often is denoted with <span class="math inline">\(H\)</span>, and here <span class="math inline">\(p\)</span> denotes proportion of the <span class="math inline">\(i\)</span>‚Äôth taxon.</li>
</ul>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-14-contents" aria-controls="callout-14" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Explanation of the Shannon index formula
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-14" class="callout-14-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>For each taxon <span class="math inline">\(i\)</span>, you multiply its proportional abundance <span class="math inline">\(p_i\)</span> by the natural log of that proportion <span class="math inline">\(\ln p_i\)</span>, and sum those values.</li>
<li>Try it out for yourself to convince yourself you get larger (negative) values for higher proportions.</li>
<li>The highest value you can achieve with <span class="math inline">\(N\)</span> taxa occurs with equal proportions</li>
<li>e.g.&nbsp;with 20 taxa, maximum diversity occurs if each has a proportion of 5%</li>
<li>Lastly, you change the sign of the result to a positive number, for ease of interpretation</li>
<li>This just makes more intuitive sense: as higher positive numbers indicates higher diversity.</li>
</ul>
</div>
</div>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<div class="cell">
<details>
<summary>Code to compute Exponential Shannon and make a barplot</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>mice <span class="sc">%&gt;%</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_filter</span>(treatment <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Vehicle"</span>, <span class="st">"AmpMetro"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_calc_diversity</span>(<span class="at">rank =</span> <span class="st">"Family"</span>, <span class="at">index =</span> <span class="st">"shannon"</span>, <span class="at">exp =</span> <span class="cn">TRUE</span>, <span class="at">varname =</span> <span class="st">"Exp_Shannon"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_arrange</span>(Exp_Shannon) <span class="sc">%&gt;%</span> </span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_mutate</span>(<span class="at">Exp_Shannon =</span> <span class="fu">formatC</span>(Exp_Shannon, <span class="at">digits =</span> <span class="dv">2</span>, <span class="at">format =</span> <span class="st">"f"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_barplot</span>(</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_level =</span> <span class="st">"Family"</span>, <span class="at">n_taxa =</span> <span class="dv">14</span>, <span class="at">label =</span> <span class="st">"Exp_Shannon"</span>, <span class="at">bar_width =</span> <span class="fl">0.8</span>,</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample_order =</span> <span class="st">"asis"</span>, <span class="at">merge_other =</span> <span class="cn">FALSE</span>, <span class="at">facet_by =</span> <span class="st">"treatment"</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="exercises_1_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<ul>
<li>The exponent of the Shannon index <span class="math inline">\(e^H\)</span> represents the number of taxa that would be present in an evenly abundant ecosystem with the same Shannon index.</li>
</ul>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Explanation of the Effective Shannon Diversity
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>The numeric value of the Shannon index itself has no intuitive meaning.</li>
<li>You can compare them, but can‚Äôt easily interpret any one number.</li>
<li>So, the concept of ‚Äúeffective numbers‚Äù of taxa is useful here.</li>
<li>If your original ecosystem was actually perfectly even, then <span class="math inline">\(e^H = N\)</span></li>
<li>Where N is the observed richness</li>
<li>The more uneven an ecosystem, the further <span class="math inline">\(e^H\)</span> will be from <span class="math inline">\(N\)</span></li>
</ul>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="comparing-diversity" class="level2">
<h2 class="anchored" data-anchor-id="comparing-diversity">Comparing Diversity</h2>
<ul>
<li>Let us now compare the diversity of our samples.</li>
<li>Perhaps we hypothesise that the mouse gut diversity differs after antibiotic treatment.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>miceShannonDf <span class="ot">&lt;-</span> mice <span class="sc">%&gt;%</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_calc_diversity</span>(<span class="at">rank =</span> <span class="st">"Family"</span>, <span class="at">index =</span> <span class="st">"shannon"</span>, <span class="at">exp =</span> <span class="cn">TRUE</span>, <span class="at">varname =</span> <span class="st">"Exp_Shannon"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">samdat_tbl</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>miceShannonDf <span class="sc">%&gt;%</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">y =</span> Exp_Shannon, <span class="at">x =</span> treatment, <span class="at">color =</span> treatment)) <span class="sc">+</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">width =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">width =</span> <span class="fl">0.1</span>), <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>, <span class="at">guide =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Effective Shannon diversity (Family)"</span>, <span class="at">x =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_1_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<ul>
<li>The mice exposed to Ampicillin (¬± Metronidazole) appear to have higher diversity at this day 7 sample!</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># A simple statistical test supports this assertion</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>miceShannonDf <span class="sc">%&gt;%</span> </span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Ampicillin =</span> <span class="fu">if_else</span>(<span class="fu">grepl</span>(<span class="at">pattern =</span> <span class="st">"Amp"</span>, <span class="at">x =</span> treatment), <span class="dv">1</span>, <span class="dv">0</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">wilcox.test</span>(<span class="at">formula =</span> Exp_Shannon <span class="sc">~</span> Ampicillin, <span class="at">data =</span> .)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum exact test

data:  Exp_Shannon by Ampicillin
W = 59, p-value = 2.919e-06
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
</div>
<ul>
<li>You can apply more complex statistical tests if you like</li>
<li>e.g.&nbsp;adjusting for covariates with linear regression, using <code>lm()</code></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>allMice <span class="sc">%&gt;%</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_filter</span>(treatment_days <span class="sc">==</span> <span class="st">"D7"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_calc_diversity</span>(<span class="at">rank =</span> <span class="st">"Family"</span>, <span class="at">index =</span> <span class="st">"shannon"</span>, <span class="at">exp =</span> <span class="cn">TRUE</span>, <span class="at">varname =</span> <span class="st">"Exp_Shannon"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">samdat_tbl</span>() <span class="sc">%&gt;%</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Ampicillin =</span> <span class="fu">if_else</span>(<span class="fu">grepl</span>(<span class="at">pattern =</span> <span class="st">"Amp"</span>, <span class="at">x =</span> treatment), <span class="dv">1</span>, <span class="dv">0</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lm</span>(<span class="at">formula =</span> Exp_Shannon <span class="sc">~</span> Ampicillin <span class="sc">+</span> virus, <span class="at">data =</span> .) <span class="sc">%&gt;%</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summary</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = Exp_Shannon ~ Ampicillin + virus, data = .)

Residuals:
    Min      1Q  Median      3Q     Max 
-5.0571 -0.5624  0.1993  1.0744  4.2906 

Coefficients:
             Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)    1.9133     0.4977   3.844 0.000287 ***
Ampicillin     2.9685     0.4844   6.128 6.76e-08 ***
virusWNV2000   1.1967     0.5233   2.287 0.025624 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 1.944 on 62 degrees of freedom
Multiple R-squared:  0.417, Adjusted R-squared:  0.3982 
F-statistic: 22.17 on 2 and 62 DF,  p-value: 5.454e-08</code></pre>
</div>
</div>
<p><strong>Final activities for part 1:</strong></p>
<ul>
<li>Try making other diversity comparisons for yourself?</li>
<li>Consider comparing ASV-level diversity?</li>
<li>Does virus exposure influence diversity?</li>
<li>What about comparing richness?</li>
<li>Practice making plots and doing simple statistical tests.</li>
</ul>
<hr>
</section>
<section id="break-time" class="level2">
<h2 class="anchored" data-anchor-id="break-time">Break Time! ‚òïÔ∏è ‚åöÔ∏è</h2>
<section id="you-want-more-about-diversity" class="level3">
<h3 class="anchored" data-anchor-id="you-want-more-about-diversity">You want more about diversity?</h3>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-16-contents" aria-controls="callout-16" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Click here for additional notes on richness and readcount
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-16" class="callout-16-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>Simple measures like Observed Richness are sensitive to what ecologists call ‚Äúsampling effort‚Äù.</li>
<li>For macroecologists, this is actually how much time/effort you spent trying to count all the organisms present in an ecosystem.</li>
<li>In our case, the amount of total reads obtained represents the sampling effort: more reads, more effort.</li>
<li>Indeed we can see that the samples with a much lower readcount have lower observed richness.</li>
<li>Furthermore, as this richness estimate is based on a sample, and not the actual ecosystem, the richness estimate actually has quantifiable uncertainty too.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>mice <span class="sc">%&gt;%</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_filter</span>(treatment <span class="sc">==</span> <span class="st">"Amp"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_calc_richness</span>(<span class="at">rank =</span> <span class="st">"Family"</span>, <span class="at">index =</span> <span class="st">"observed"</span>, <span class="at">varname =</span> <span class="st">"N families"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_barplot</span>(</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_level =</span> <span class="st">"Family"</span>, <span class="at">n_taxa =</span> <span class="dv">12</span>, <span class="at">label =</span> <span class="st">"N families"</span>, <span class="at">bar_width =</span> <span class="fl">0.7</span>,</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample_order =</span> <span class="st">"asis"</span>, <span class="at">merge_other =</span> <span class="cn">FALSE</span>, <span class="at">tax_transform_for_plot =</span> <span class="st">"identity"</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_1_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>mice <span class="sc">%&gt;%</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_names2rank</span>(<span class="st">"ASV"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_calc_richness</span>(<span class="at">rank =</span> <span class="st">"ASV"</span>, <span class="at">index =</span> <span class="st">"observed"</span>, <span class="at">varname =</span> <span class="st">"ASVs"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_mutate</span>(<span class="at">readcount =</span> <span class="fu">sample_sums</span>(mice)) <span class="sc">%&gt;%</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">samdat_tbl</span>() <span class="sc">%&gt;%</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(readcount, ASVs)) <span class="sc">+</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>, <span class="at">size =</span> <span class="fl">2.5</span>) <span class="sc">+</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="dv">14</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_1_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="480"></p>
</div>
</div>
<p><strong>What to do?</strong></p>
<ol type="1">
<li><strong>Simple solution</strong>: Ignore the problem. Whilst you can‚Äôt interpret the richness of any individual sample as being correct, it is still usually valid to compare richness across groups of samples, as the readcount variation is only random noise, and should be uncorrelated with your grouping variable (but do check this).</li>
<li><strong>Harder solution</strong>: Explore more rigorous methods like <code>breakaway</code> by Amy Willis and team. <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2019.02407/full" class="uri">https://www.frontiersin.org/articles/10.3389/fmicb.2019.02407/full</a></li>
</ol>
</div>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-17-contents" aria-controls="callout-17" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Click here for an extension activity with an IBD dataset.
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-17" class="callout-17-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><strong>This is an extension exercise, for anyone who is moving very quickly.</strong></p>
<p><strong>Inflammatory Bowel Disease study</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>ibd <span class="ot">&lt;-</span> corncob<span class="sc">::</span>ibd_phylo <span class="sc">%&gt;%</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_mutate</span>(<span class="at">Species =</span> <span class="cn">NULL</span>) <span class="sc">%&gt;%</span> <span class="co"># ibd_phylo Species column was blank -&gt; deleted</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_mutate</span>(<span class="at">disease =</span> ibd <span class="sc">==</span> <span class="st">"ibd"</span>, <span class="at">ibd =</span> <span class="cn">NULL</span>) <span class="co"># adds disease state indicator variable</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><code>ibd</code> is another phyloseq object containing 16S rRNA gene amplicon sequencing data</li>
<li>It is from a 2012 study of Inflammatory Bowel Disease in children and young adults</li>
<li>It‚Äôs ‚Äúold‚Äù data: they used 454 Pyrosequencing, and clustered the raw sequences into ‚ÄúOTUs‚Äù</li>
<li>Have a look at the data, like we did before for the <code>mice</code> dataset</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>ibd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 36349 taxa and 91 samples ]
sample_data() Sample Data:       [ 91 samples by 15 sample variables ]
tax_table()   Taxonomy Table:    [ 36349 taxa by 6 taxonomic ranks ]</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>You can perform alpha diversity analysis:</li>
<li>Try comparing the alpha diversity of the IBD patients against the healthy controls.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="session-info" class="level2">
<h2 class="anchored" data-anchor-id="session-info">Session info</h2>
<p><code>session_info</code> records your package versions etc. This is useful for debugging / reproducing analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">session_info</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>‚îÄ Session info ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ
 setting  value
 version  R version 4.3.0 (2023-04-21)
 os       macOS Ventura 13.3.1
 system   aarch64, darwin20
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Europe/Prague
 date     2023-05-23
 pandoc   3.0.1 @ /opt/homebrew/bin/ (via rmarkdown)

‚îÄ Packages ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ
 package          * version   date (UTC) lib source
 ade4               1.7-22    2023-02-06 [1] CRAN (R 4.3.0)
 ape                5.7-1     2023-03-13 [1] CRAN (R 4.3.0)
 Biobase            2.60.0    2023-04-25 [1] RSPM (R 4.3.0)
 BiocGenerics       0.46.0    2023-04-25 [1] RSPM (R 4.3.0)
 biomformat         1.28.0    2023-04-25 [1] RSPM (R 4.3.0)
 Biostrings         2.68.0    2023-04-25 [1] RSPM (R 4.3.0)
 bitops             1.0-7     2021-04-24 [1] CRAN (R 4.3.0)
 ca                 0.71.1    2020-01-24 [1] CRAN (R 4.3.0)
 cachem             1.0.8     2023-05-01 [1] CRAN (R 4.3.0)
 callr              3.7.3     2022-11-02 [1] CRAN (R 4.3.0)
 cli                3.6.1     2023-03-23 [1] CRAN (R 4.3.0)
 cluster            2.1.4     2022-08-22 [1] CRAN (R 4.3.0)
 codetools          0.2-19    2023-02-01 [1] CRAN (R 4.3.0)
 colorspace         2.1-0     2023-01-23 [1] CRAN (R 4.3.0)
 corncob            0.3.1     2022-12-05 [1] CRAN (R 4.3.0)
 crayon             1.5.2     2022-09-29 [1] CRAN (R 4.3.0)
 data.table         1.14.8    2023-02-17 [1] CRAN (R 4.3.0)
 devtools           2.4.5     2022-10-11 [1] CRAN (R 4.3.0)
 digest             0.6.31    2022-12-11 [1] CRAN (R 4.3.0)
 dplyr            * 1.1.2     2023-04-20 [1] CRAN (R 4.3.0)
 ellipsis           0.3.2     2021-04-29 [1] CRAN (R 4.3.0)
 evaluate           0.21      2023-05-05 [1] CRAN (R 4.3.0)
 fansi              1.0.4     2023-01-22 [1] CRAN (R 4.3.0)
 farver             2.1.1     2022-07-06 [1] CRAN (R 4.3.0)
 fastmap            1.1.1     2023-02-24 [1] CRAN (R 4.3.0)
 forcats          * 1.0.0     2023-01-29 [1] CRAN (R 4.3.0)
 foreach            1.5.2     2022-02-02 [1] CRAN (R 4.3.0)
 fs                 1.6.2     2023-04-25 [1] CRAN (R 4.3.0)
 generics           0.1.3     2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb       1.36.0    2023-04-25 [1] RSPM (R 4.3.0)
 GenomeInfoDbData   1.2.10    2023-05-12 [1] RSPM (R 4.3.0)
 ggplot2          * 3.4.2     2023-04-03 [1] CRAN (R 4.3.0)
 glue               1.6.2     2022-02-24 [1] CRAN (R 4.3.0)
 gtable             0.3.3     2023-03-21 [1] CRAN (R 4.3.0)
 here               1.0.1     2020-12-13 [1] CRAN (R 4.3.0)
 hms                1.1.3     2023-03-21 [1] CRAN (R 4.3.0)
 htmltools          0.5.5     2023-03-23 [1] CRAN (R 4.3.0)
 htmlwidgets        1.6.2     2023-03-17 [1] CRAN (R 4.3.0)
 httpuv             1.6.9     2023-02-14 [1] CRAN (R 4.3.0)
 igraph             1.4.2     2023-04-07 [1] CRAN (R 4.3.0)
 IRanges            2.34.0    2023-04-25 [1] RSPM (R 4.3.0)
 iterators          1.0.14    2022-02-05 [1] CRAN (R 4.3.0)
 jsonlite           1.8.4     2022-12-06 [1] CRAN (R 4.3.0)
 knitr              1.42      2023-01-25 [1] CRAN (R 4.3.0)
 labeling           0.4.2     2020-10-20 [1] CRAN (R 4.3.0)
 later              1.3.1     2023-05-02 [1] CRAN (R 4.3.0)
 lattice            0.21-8    2023-04-05 [1] CRAN (R 4.3.0)
 lifecycle          1.0.3     2022-10-07 [1] CRAN (R 4.3.0)
 lubridate        * 1.9.2     2023-02-10 [1] CRAN (R 4.3.0)
 magrittr           2.0.3     2022-03-30 [1] CRAN (R 4.3.0)
 MASS               7.3-60    2023-05-04 [1] CRAN (R 4.3.0)
 Matrix             1.5-4     2023-04-04 [1] CRAN (R 4.3.0)
 memoise            2.0.1     2021-11-26 [1] CRAN (R 4.3.0)
 mgcv               1.8-42    2023-03-02 [1] CRAN (R 4.3.0)
 microbiome         1.22.0    2023-04-25 [1] RSPM (R 4.3.0)
 microViz         * 0.10.10   2023-05-19 [1] Github (david-barnett/microViz@2a021f2)
 mime               0.12      2021-09-28 [1] CRAN (R 4.3.0)
 miniUI             0.1.1.1   2018-05-18 [1] CRAN (R 4.3.0)
 multtest           2.56.0    2023-04-25 [1] RSPM (R 4.3.0)
 munsell            0.5.0     2018-06-12 [1] CRAN (R 4.3.0)
 nlme               3.1-162   2023-01-31 [1] CRAN (R 4.3.0)
 patchwork        * 1.1.2     2022-08-19 [1] CRAN (R 4.3.0)
 permute            0.9-7     2022-01-27 [1] CRAN (R 4.3.0)
 phyloseq         * 1.44.0    2023-04-25 [1] RSPM (R 4.3.0)
 pillar             1.9.0     2023-03-22 [1] CRAN (R 4.3.0)
 pkgbuild           1.4.0     2022-11-27 [1] CRAN (R 4.3.0)
 pkgconfig          2.0.3     2019-09-22 [1] CRAN (R 4.3.0)
 pkgload            1.3.2     2022-11-16 [1] CRAN (R 4.3.0)
 plyr               1.8.8     2022-11-11 [1] CRAN (R 4.3.0)
 prettyunits        1.1.1     2020-01-24 [1] CRAN (R 4.3.0)
 processx           3.8.1     2023-04-18 [1] CRAN (R 4.3.0)
 profvis            0.3.8     2023-05-02 [1] CRAN (R 4.3.0)
 promises           1.2.0.1   2021-02-11 [1] CRAN (R 4.3.0)
 ps                 1.7.5     2023-04-18 [1] CRAN (R 4.3.0)
 purrr            * 1.0.1     2023-01-10 [1] CRAN (R 4.3.0)
 R6                 2.5.1     2021-08-19 [1] CRAN (R 4.3.0)
 ragg               1.2.5     2023-01-12 [1] CRAN (R 4.3.0)
 RColorBrewer       1.1-3     2022-04-03 [1] CRAN (R 4.3.0)
 Rcpp               1.0.10    2023-01-22 [1] CRAN (R 4.3.0)
 RCurl              1.98-1.12 2023-03-27 [1] CRAN (R 4.3.0)
 readr            * 2.1.4     2023-02-10 [1] CRAN (R 4.3.0)
 registry           0.5-1     2019-03-05 [1] CRAN (R 4.3.0)
 remotes            2.4.2     2021-11-30 [1] CRAN (R 4.3.0)
 reshape2           1.4.4     2020-04-09 [1] CRAN (R 4.3.0)
 rhdf5              2.44.0    2023-04-25 [1] RSPM (R 4.3.0)
 rhdf5filters       1.12.1    2023-04-30 [1] RSPM (R 4.3.0)
 Rhdf5lib           1.22.0    2023-04-25 [1] RSPM (R 4.3.0)
 rlang              1.1.1     2023-04-28 [1] CRAN (R 4.3.0)
 rmarkdown          2.21      2023-03-26 [1] CRAN (R 4.3.0)
 rprojroot          2.0.3     2022-04-02 [1] CRAN (R 4.3.0)
 rstudioapi         0.14      2022-08-22 [1] CRAN (R 4.3.0)
 Rtsne              0.16      2022-04-17 [1] CRAN (R 4.3.0)
 S4Vectors          0.38.1    2023-05-02 [1] RSPM (R 4.3.0)
 scales             1.2.1     2022-08-20 [1] CRAN (R 4.3.0)
 seriation          1.4.2     2023-03-08 [1] CRAN (R 4.3.0)
 sessioninfo        1.2.2     2021-12-06 [1] CRAN (R 4.3.0)
 shiny            * 1.7.4     2022-12-15 [1] CRAN (R 4.3.0)
 stringi            1.7.12    2023-01-11 [1] CRAN (R 4.3.0)
 stringr          * 1.5.0     2022-12-02 [1] CRAN (R 4.3.0)
 survival           3.5-5     2023-03-12 [1] CRAN (R 4.3.0)
 systemfonts        1.0.4     2022-02-11 [1] CRAN (R 4.3.0)
 textshaping        0.3.6     2021-10-13 [1] CRAN (R 4.3.0)
 tibble           * 3.2.1     2023-03-20 [1] CRAN (R 4.3.0)
 tidyr            * 1.3.0     2023-01-24 [1] CRAN (R 4.3.0)
 tidyselect         1.2.0     2022-10-10 [1] CRAN (R 4.3.0)
 tidyverse        * 2.0.0     2023-02-22 [1] CRAN (R 4.3.0)
 timechange         0.2.0     2023-01-11 [1] CRAN (R 4.3.0)
 TSP                1.2-4     2023-04-04 [1] CRAN (R 4.3.0)
 tzdb               0.3.0     2022-03-28 [1] CRAN (R 4.3.0)
 urlchecker         1.0.1     2021-11-30 [1] CRAN (R 4.3.0)
 usethis            2.1.6     2022-05-25 [1] CRAN (R 4.3.0)
 utf8               1.2.3     2023-01-31 [1] CRAN (R 4.3.0)
 vctrs              0.6.2     2023-04-19 [1] CRAN (R 4.3.0)
 vegan              2.6-4     2022-10-11 [1] CRAN (R 4.3.0)
 withr              2.5.0     2022-03-03 [1] CRAN (R 4.3.0)
 xfun               0.39      2023-04-20 [1] CRAN (R 4.3.0)
 xtable             1.8-4     2019-04-21 [1] CRAN (R 4.3.0)
 XVector            0.40.0    2023-04-25 [1] RSPM (R 4.3.0)
 yaml               2.3.7     2023-01-23 [1] CRAN (R 4.3.0)
 zlibbioc           1.46.0    2023-04-25 [1] RSPM (R 4.3.0)

 [1] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>