<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.313">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="David Barnett">
<meta name="dcterms.date" content="2023-05-24">

<title>Evomics Workshop part 2</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="exercises_2_files/libs/clipboard/clipboard.min.js"></script>
<script src="exercises_2_files/libs/quarto-html/quarto.js"></script>
<script src="exercises_2_files/libs/quarto-html/popper.min.js"></script>
<script src="exercises_2_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="exercises_2_files/libs/quarto-html/anchor.min.js"></script>
<link href="exercises_2_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="exercises_2_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="exercises_2_files/libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="exercises_2_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="exercises_2_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="exercises_2_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="exercises_2_files/libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#welcome-to-part-2" id="toc-welcome-to-part-2" class="nav-link active" data-scroll-target="#welcome-to-part-2">Welcome to Part 2 üëã</a>
  <ul>
  <li><a href="#further-resources" id="toc-further-resources" class="nav-link" data-scroll-target="#further-resources">Further resources</a></li>
  </ul></li>
  <li><a href="#setup" id="toc-setup" class="nav-link" data-scroll-target="#setup">Setup ‚öôÔ∏è</a></li>
  <li><a href="#dissimilarity" id="toc-dissimilarity" class="nav-link" data-scroll-target="#dissimilarity">Dissimilarity üí© ‚ÜîÔ∏é üí© ?</a>
  <ul>
  <li><a href="#dissimilarity-measures" id="toc-dissimilarity-measures" class="nav-link" data-scroll-target="#dissimilarity-measures">Dissimilarity measures</a></li>
  </ul></li>
  <li><a href="#ordination" id="toc-ordination" class="nav-link" data-scroll-target="#ordination">Ordination</a>
  <ul>
  <li><a href="#pcoa" id="toc-pcoa" class="nav-link" data-scroll-target="#pcoa">PCoA</a></li>
  </ul></li>
  <li><a href="#interactive-ordination" id="toc-interactive-ordination" class="nav-link" data-scroll-target="#interactive-ordination">Interactive ordination!</a></li>
  <li><a href="#permanova" id="toc-permanova" class="nav-link" data-scroll-target="#permanova">PERMANOVA</a>
  <ul>
  <li><a href="#reporting-pcoa-and-permanova-methods" id="toc-reporting-pcoa-and-permanova-methods" class="nav-link" data-scroll-target="#reporting-pcoa-and-permanova-methods">Reporting PCoA and PERMANOVA methods</a></li>
  <li><a href="#covariate-adjusted-permanova" id="toc-covariate-adjusted-permanova" class="nav-link" data-scroll-target="#covariate-adjusted-permanova">Covariate-adjusted PERMANOVA</a></li>
  </ul></li>
  <li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca">PCA</a></li>
  <li><a href="#log-transformations-and-clr" id="toc-log-transformations-and-clr" class="nav-link" data-scroll-target="#log-transformations-and-clr">Log transformations, and CLR</a>
  <ul>
  <li><a href="#centered-log-ratio-transformation" id="toc-centered-log-ratio-transformation" class="nav-link" data-scroll-target="#centered-log-ratio-transformation">Centered Log Ratio transformation:</a></li>
  <li><a href="#pca-on-clr-transformed-taxa" id="toc-pca-on-clr-transformed-taxa" class="nav-link" data-scroll-target="#pca-on-clr-transformed-taxa">PCA on CLR-transformed taxa</a>
  <ul class="collapse">
  <li><a href="#so-why-is-pca-interesting-for-us" id="toc-so-why-is-pca-interesting-for-us" class="nav-link" data-scroll-target="#so-why-is-pca-interesting-for-us">So why is PCA interesting for us?</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#taxon-stats" id="toc-taxon-stats" class="nav-link" data-scroll-target="#taxon-stats">Taxon stats</a>
  <ul>
  <li><a href="#model-one-taxon" id="toc-model-one-taxon" class="nav-link" data-scroll-target="#model-one-taxon">Model one taxon</a>
  <ul class="collapse">
  <li><a href="#many-da-methods" id="toc-many-da-methods" class="nav-link" data-scroll-target="#many-da-methods">Many DA methods</a></li>
  <li><a href="#so-what-to-do" id="toc-so-what-to-do" class="nav-link" data-scroll-target="#so-what-to-do">So, what to do?</a></li>
  </ul></li>
  <li><a href="#model-all-the-taxa" id="toc-model-all-the-taxa" class="nav-link" data-scroll-target="#model-all-the-taxa">Model all the taxa!</a>
  <ul class="collapse">
  <li><a href="#getting-stats-from-the-models" id="toc-getting-stats-from-the-models" class="nav-link" data-scroll-target="#getting-stats-from-the-models">Getting stats from the models</a></li>
  <li><a href="#adjusting-p-values" id="toc-adjusting-p-values" class="nav-link" data-scroll-target="#adjusting-p-values">Adjusting p values</a></li>
  </ul></li>
  <li><a href="#plot-all-the-taxatree_stats" id="toc-plot-all-the-taxatree_stats" class="nav-link" data-scroll-target="#plot-all-the-taxatree_stats">Plot all the taxatree_stats!</a>
  <ul class="collapse">
  <li><a href="#taxatree-key" id="toc-taxatree-key" class="nav-link" data-scroll-target="#taxatree-key">Taxatree Key</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session info</a></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Evomics Workshop part 2</h1>
<p class="subtitle lead">Microbiome analysis with microViz</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>David Barnett </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">May 24, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="welcome-to-part-2" class="level2">
<h2 class="anchored" data-anchor-id="welcome-to-part-2">Welcome to Part 2 üëã</h2>
<ul>
<li>Bar charts and diversity were relatively straightforward topics, conceptually.</li>
<li>Let‚Äôs look at some more things we can do with microbiome composition data!</li>
</ul>
<p><strong>Topics for part 2:</strong></p>
<ol type="1">
<li>Dissimilarity measures</li>
<li>Ordination: PCoA &amp; PCA (with interactive data exploration)</li>
<li>Differential abundance testing</li>
</ol>
<section id="further-resources" class="level3">
<h3 class="anchored" data-anchor-id="further-resources">Further resources</h3>
<p>Refer to the <a href="https://david-barnett.github.io/microViz/">microViz website</a> to see help pages for every function (as well as further tutorials).</p>
<hr>
</section>
</section>
<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup ‚öôÔ∏è</h2>
<p>Load the R packages we will be using:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(seriation)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microViz)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(shiny)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="dissimilarity" class="level2">
<h2 class="anchored" data-anchor-id="dissimilarity">Dissimilarity üí© ‚ÜîÔ∏é üí© ?</h2>
<ul>
<li><strong>Ecosystem Diversity</strong> is sometimes referred to as ‚Äúalpha‚Äù diversity or ‚Äúwithin-sample‚Äù diversity</li>
<li>Now we‚Äôre going to look at <strong>Dissimilarity</strong> between ecosystems</li>
<li>Sometimes this is confusingly referred to as ‚Äúbeta diversity‚Äù analysis, or ‚Äúbetween-sample‚Äù diversity</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"shao19"</span>) <span class="co"># this example data is built in to microViz</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>For this part we‚Äôre going to use an infant gut microbiome shotgun metagenomics dataset.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>shao19</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 819 taxa and 1644 samples ]
sample_data() Sample Data:       [ 1644 samples by 11 sample variables ]
tax_table()   Taxonomy Table:    [ 819 taxa by 6 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 819 tips and 818 internal nodes ]</code></pre>
</div>
</div>
<ul>
<li>Do you remember how to examine a phyloseq object?</li>
<li>Look at the rank names, sample data variables etc.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>?shao19 <span class="co"># in the console, for info on the data source</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
A foreword on filtering
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>You should check if any of your samples have a surprisingly low total number of (classified) reads.</li>
<li>This can suggest that something went wrong in the lab (or during sample collection)</li>
<li>The data from these samples might be unreliable.</li>
<li>You might already do a check for total reads and remove poor quality samples during the fastq file processing.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>shao19 <span class="sc">%&gt;%</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_mutate</span>(<span class="at">reads =</span> <span class="fu">sample_sums</span>(shao19)) <span class="sc">%&gt;%</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">samdat_tbl</span>() <span class="sc">%&gt;%</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> reads)) <span class="sc">+</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_freqpoly</span>(<span class="at">bins =</span> <span class="dv">500</span>) <span class="sc">+</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rug</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>(<span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">label_number</span>()) <span class="sc">+</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Number of classified reads"</span>, <span class="at">y =</span> <span class="cn">NULL</span>) <span class="sc">+</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p><strong>How many is enough? There is no easy answer!</strong></p>
<p>These samples have great depth. There are a few with much less reads than the rest, and a few with under a million. You might consider dropping the samples with under a million reads, to see if it affects your results, but in this case we won‚Äôt.</p>
<p>But 100,000 is still a lot, compared to what older sequencing machines produced: 1000 reads might have been considered very good. So look at the distribution for your data, in case there are obvious outliers, and look at recent papers using a similar sequencing technique for what kind of threshold they used.</p>
<p>There might also be relevant information for the type of sequencer you used on e.g. Illumina website. e.g.&nbsp;for this type of sequencing Illumina suggests you should expect at least a million reads (and this is good for RNA seq analyses). <a href="https://support.illumina.com/bulletins/2017/04/considerations-for-rna-seq-read-length-and-coverage-.html" class="uri">https://support.illumina.com/bulletins/2017/04/considerations-for-rna-seq-read-length-and-coverage-.html</a></p>
</div>
</div>
</div>
<hr>
<section id="dissimilarity-measures" class="level3">
<h3 class="anchored" data-anchor-id="dissimilarity-measures">Dissimilarity measures</h3>
<ul>
<li>What are we doing?</li>
<li>Calculating the dissimilarities between pairs of microbiome samples</li>
<li>We talked about these commonly-used dissimilarity measures in the lecture.
<ul>
<li><strong>Binary Jaccard</strong> - presence-absence</li>
<li><strong>Bray-Curtis</strong> - abundance-weighted</li>
<li><strong>UniFrac</strong> distances (unweighted / weighted / generalised)</li>
</ul></li>
<li>To simplify and speed up the analyses, we‚Äôre going to take a smaller part of the dataset.</li>
<li>We‚Äôll only look at the 300 infant fecal samples from 4 days of age.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>infants <span class="ot">&lt;-</span> shao19 <span class="sc">%&gt;%</span> <span class="fu">ps_filter</span>(family_role <span class="sc">==</span> <span class="st">"child"</span>, infant_age <span class="sc">==</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>We‚Äôre going to filter out rare taxa quite strictly, for similar reasons</li>
<li>But we won‚Äôt overwrite our smaller dataset: we‚Äôll do the filtering per analysis</li>
<li>You will find out more about the how and why of taxa filtering later</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="fl">2.5</span> <span class="sc">/</span> <span class="dv">100</span>) <span class="sc">%&gt;%</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_agg</span>(<span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="st">"binary"</span>) <span class="sc">%&gt;%</span> <span class="co"># converts counts to absence/presence: 0/1</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist_calc</span>(<span class="at">dist =</span> <span class="st">"jaccard"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Proportional min_prevalence given: 0.025 --&gt; min 8/306 samples.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>psExtra object - a phyloseq object with extra slots:

phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 35 taxa and 306 samples ]
sample_data() Sample Data:       [ 306 samples by 11 sample variables ]
tax_table()   Taxonomy Table:    [ 35 taxa by 5 taxonomic ranks ]

otu_get(counts = TRUE)       [ 35 taxa and 306 samples ]

psExtra info:
tax_agg = "genus" tax_trans = "binary" 

jaccard distance matrix of size 306 
0.6666667 0.7333333 0.9375 0.8125 0.6428571 ...</code></pre>
</div>
</div>
<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
‚ÄúBinary‚Äù Jaccard <code>010101</code>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>Remember to run a ‚Äúbinary‚Äù transform on your data before computing ‚Äújaccard‚Äù distance.</li>
<li>There is a quantitative form of the Jaccard distance, which is the default behaviour!</li>
<li>But the qualitative (presence/absence) version is mostly used in microbial ecology.</li>
<li>If you want an abundance-weighted ecological dissimilarity, use Bray-Curtis!</li>
</ul>
</div>
</div>
</div>
<ul>
<li>We now have our pairwise dissimilarities! üéâ</li>
<li>A distance matrix is attached as an extra part on the original phyloseq object</li>
</ul>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Dissimilarity or distance?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>These terms are often used interchangeably</li>
<li>You will find dissimilarities in a distance matrix</li>
<li>But if you want to be pedantic a true ‚Äúdistance metric‚Äù d, must satisfy 3 properties:
<ol type="1">
<li>Identity of indiscernibles: For any samples x and y, d(x, y) = 0 if and only if x = y</li>
<li>Symmetry: For any samples x and y, d(x, y) = d(y, x)</li>
<li>Triangle inequality: For any samples x, y, and z, d(x, z) ‚â§ d(x, y) + d(y, z).</li>
</ol></li>
<li><ol start="3" type="1">
<li>can be interpreted as: ‚Äúthe direct path between two points must be at least as short as any detour‚Äù</li>
</ol></li>
<li>This is not true for e.g.&nbsp;Bray-Curtis, but in practice it is very rarely problematic.</li>
</ul>
</div>
</div>
</div>
<ul>
<li>The object is now class ‚ÄúpsExtra‚Äù (created by microViz)</li>
<li>A psExtra also stores info about the aggregation and transformations you performed</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>distances <span class="ot">&lt;-</span> infants <span class="sc">%&gt;%</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="fl">2.5</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_agg</span>(<span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="st">"binary"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist_calc</span>(<span class="at">dist =</span> <span class="st">"jaccard"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist_get</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can extract the distance matrix with dist_get.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.matrix</span>(distances)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            B01089_ba_4 B01190_ba_4 B01194_ba_4 B01196_ba_4 B01235_ba_4
B01089_ba_4   0.0000000   0.6666667   0.7333333   0.9375000   0.8125000
B01190_ba_4   0.6666667   0.0000000   0.7500000   0.9166667   0.8461538
B01194_ba_4   0.7333333   0.7500000   0.0000000   0.4615385   0.3076923
B01196_ba_4   0.9375000   0.9166667   0.4615385   0.0000000   0.4615385
B01235_ba_4   0.8125000   0.8461538   0.3076923   0.4615385   0.0000000</code></pre>
</div>
</div>
<p>The Binary Jaccard dissimilarities range between 0 (identical) and 1 (no shared genera).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">range</span>(distances)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0 1</code></pre>
</div>
</div>
<hr>
</section>
</section>
<section id="ordination" class="level2">
<h2 class="anchored" data-anchor-id="ordination">Ordination</h2>
<ul>
<li>What can we do with these dissimilarities? ü§î</li>
<li>We can make an ordination! üí°</li>
<li>Conceptually, ordination refers to a process of ordering things (in our case: samples).</li>
<li>Similar samples are placed closer to each other, and dissimilar samples are placed further away.</li>
</ul>
<section id="pcoa" class="level3">
<h3 class="anchored" data-anchor-id="pcoa">PCoA</h3>
<p>Principal Co-ordinates Analysis is one kind of ordination.</p>
<ul>
<li>PCoA takes a distance matrix and finds new dimensions (a co-ordinate system, if you like)</li>
<li>The new dimensions are created with the aim to preserve the original distances between samples</li>
<li>And to capture the majority of this distance information in the first dimensions</li>
<li>This makes it easier to visualize the patterns in your data, in 2D or 3D plots üëÄ</li>
</ul>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
For more info, see ‚ÄúGUSTAME‚Äù
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>There is helpful info about ordination methods including PCoA on the GUide to STatistical Analysis in Microbial Ecology website (GUSTA ME). <a href="https://sites.google.com/site/mb3gustame/dissimilarity-based-methods/principal-coordinates-analysis" class="uri">https://sites.google.com/site/mb3gustame/dissimilarity-based-methods/principal-coordinates-analysis</a></li>
<li>This website covers a lot of other topics too, which may be interesting for you to read at a later date if you‚Äôll work on microbiome analysis.</li>
</ul>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="fl">2.5</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="at">trans =</span> <span class="st">"identity"</span>, <span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist_calc</span>(<span class="at">dist =</span> <span class="st">"bray"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_calc</span>(<span class="at">method =</span> <span class="st">"PCoA"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_plot</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>(<span class="dv">12</span>) <span class="sc">+</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>(<span class="fl">0.7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>To get a little insight into what has happened here, we can colour each sample according to its dominant (most abundant) genus.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="fl">2.5</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_calc_dominant</span>(<span class="at">rank =</span> <span class="st">"genus"</span>, <span class="at">none =</span> <span class="st">"Mixed"</span>, <span class="at">other =</span> <span class="st">"Other"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="at">trans =</span> <span class="st">"identity"</span>, <span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist_calc</span>(<span class="at">dist =</span> <span class="st">"bray"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_calc</span>(<span class="at">method =</span> <span class="st">"PCoA"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_plot</span>(<span class="at">color =</span> <span class="st">"dominant_genus"</span>, <span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">name =</span> <span class="st">"Dominant Genus"</span>, <span class="at">palette =</span> <span class="st">"Dark2"</span>) <span class="sc">+</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>(<span class="dv">12</span>) <span class="sc">+</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>(<span class="fl">0.7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="576"></p>
</div>
</div>
</section>
</section>
<section id="interactive-ordination" class="level2">
<h2 class="anchored" data-anchor-id="interactive-ordination">Interactive ordination!</h2>
<p><code>microViz</code> provides a Shiny app <code>ord_explore</code> to interactively create and explore PCoA plots and other ordinations. See the code below to get started. But read the instructions first.</p>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
<strong>Instructions:</strong> a few things to try out
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>Colour the samples using the variables in the sample data</li>
<li>Select a few samples to view their composition on barplots!</li>
<li>Change some ordination options:
<ul>
<li>Different rank of taxonomic aggregation</li>
<li>Different distances we‚Äôve discussed</li>
</ul></li>
<li>Copy the automatically generated code
<ul>
<li>Exit the app (press escape or click red button in R console!)</li>
<li>Paste and run the code to recreate the ordination plot</li>
<li>Customise the plot: change colour scheme, title, etc.</li>
</ul></li>
<li>Launch the app again with a different subset of the data
<ul>
<li>Practice using <code>ps_filter</code> etc.</li>
<li>e.g.&nbsp;plot the data of the mothers‚Äô gut microbiomes!</li>
<li>compute and colour points by an alpha diversity measure?</li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="callout-warning callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
<strong>Beware: some important notes on interactive analysis</strong>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>UniFrac distances can be quite slow (over a minute) to calculate!
<ul>
<li>Filter to fewer samples and fewer taxa to speed it up (Before launching the app)</li>
</ul></li>
<li>There are many distances available, feel free to try out ones we haven‚Äôt talked about
<ul>
<li><strong>BUT</strong>:
<ul>
<li>You shouldn‚Äôt use a distance that you don‚Äôt understand in your actual work, even if the plot looks nice! üòâ</li>
<li>A few of the distances might not work‚Ä¶
<ul>
<li>They are mostly implemented in the package <code>vegan</code> and I haven‚Äôt tested them all</li>
<li>Errors will appear in the RStudio R console</li>
<li>You can report to me any distances that don‚Äôt work (if you‚Äôre feeling helpful! üòá)</li>
</ul></li>
</ul></li>
</ul></li>
<li>There are other ordination methods available in <code>ord_explore</code>
<ul>
<li>Try out PCA, principal <strong>components</strong> analysis, which does NOT use distances</li>
<li>We will not cover constrained and conditioned ordinations</li>
<li>If you are interested in e.g.&nbsp;RDA, you can look this up later</li>
<li>See the <a href="https://sites.google.com/site/mb3gustame/constrained-analyses/rda">Guide to Statistical Analysis in Microbial Ecology</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fire up the shiny app</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co"># run these lines in your console (don't keep in script/notebook)</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="fl">2.5</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate new sample variables with dominant taxon (optional)</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_calc_dominant</span>(<span class="at">rank =</span> <span class="st">"genus"</span>, <span class="at">none =</span> <span class="st">"Mixed"</span>, <span class="at">other =</span> <span class="st">"Other"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># launch a Shiny app in your web browser!</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_explore</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Again, with different options</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Run these lines in your console</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>shao19 <span class="sc">%&gt;%</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_filter</span>(family_role <span class="sc">==</span> <span class="st">"mother"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="fl">2.5</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculate a few sample variables for interest (optional)</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_calc_dominant</span>(<span class="at">rank =</span> <span class="st">"genus"</span>, <span class="at">none =</span> <span class="st">"Mixed"</span>, <span class="at">other =</span> <span class="st">"Other"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_calc_diversity</span>(<span class="at">rank =</span> <span class="st">"genus"</span>, <span class="at">index =</span> <span class="st">"shannon"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_calc_richness</span>(<span class="at">rank =</span> <span class="st">"genus"</span>, <span class="at">index =</span> <span class="st">"observed"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># launch a Shiny app in your web browser!</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_explore</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr>
</section>
<section id="permanova" class="level2">
<h2 class="anchored" data-anchor-id="permanova">PERMANOVA</h2>
<p>Permutational multivariate analysis of variance.</p>
<ul>
<li>ANOVA - analysis of variance (statistical modelling approach)</li>
<li>Multivariate - more than one dependent variable (multiple taxa!)</li>
<li>Permutational - statistical significance estimates obtained by shuffling the data many times</li>
</ul>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
For more details on PERMANOVA
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>See this excellent book chapter by Marti Anderson on PERMANOVA: <a href="https://onlinelibrary.wiley.com/doi/full/10.1002/9781118445112.stat07841" class="uri">https://onlinelibrary.wiley.com/doi/full/10.1002/9781118445112.stat07841</a></li>
<li>Sometimes PERMANOVA is called NP-MANOVA (non-parametric MANOVA)</li>
<li>e.g.&nbsp;on the GUide to STatistical Analysis in Microbial Ecology <a href="https://sites.google.com/site/mb3gustame/hypothesis-tests/manova/npmanova">website</a>.</li>
</ul>
</div>
</div>
</div>
<ul>
<li><strong>TLDR:</strong> Are those groups on the PCoA actually different??</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="fl">2.5</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_agg</span>(<span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist_calc</span>(<span class="at">dist =</span> <span class="st">"bray"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_calc</span>(<span class="at">method =</span> <span class="st">"PCoA"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_plot</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">"birth_mode"</span>) <span class="sc">+</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>(<span class="dv">12</span>) <span class="sc">+</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>(<span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_ellipse</span>(<span class="fu">aes</span>(<span class="at">color =</span> birth_mode)) <span class="sc">+</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="fl">2.5</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_agg</span>(<span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist_calc</span>(<span class="at">dist =</span> <span class="st">"bray"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist_permanova</span>(<span class="at">variables =</span> <span class="st">"birth_mode"</span>, <span class="at">n_perms =</span> <span class="dv">99</span>, <span class="at">seed =</span> <span class="dv">123</span>) <span class="sc">%&gt;%</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">perm_get</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>2023-05-23 20:39:12.303636 - Starting PERMANOVA with 99 perms with 1 processes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2023-05-23 20:39:12.396654 - Finished PERMANOVA</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for adonis under reduced model
Marginal effects of terms
Permutation: free
Number of permutations: 99

vegan::adonis2(formula = formula, data = metadata, permutations = n_perms, by = by, parallel = parall)
            Df SumOfSqs      R2      F Pr(&gt;F)   
birth_mode   1   13.790 0.12366 42.898   0.01 **
Residual   304   97.727 0.87634                 
Total      305  111.518 1.00000                 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use more permutations for a more reliable p.value in your real work (slower)</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Set a random seed number for reproducibility of this stochastic method</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can see from the model output that the p value, Pr(&gt;F) is below 0.05. So there is good statistical evidence that the bacterial gut microbiota composition of C-section delivered infants has a different composition than vaginally delivered infants at 4 days of age.</p>
<section id="reporting-pcoa-and-permanova-methods" class="level3">
<h3 class="anchored" data-anchor-id="reporting-pcoa-and-permanova-methods">Reporting PCoA and PERMANOVA methods</h3>
<ul>
<li>Your methodological choices matter, you should report what you did:
<ul>
<li>any relevant rare taxon filtering thresholds</li>
<li>the taxonomic rank of aggregation</li>
<li>the dissimilarity measure used to compute the pairwise distances</li>
</ul></li>
</ul>
<p>It‚Äôs probably a good idea to decide on a couple of appropriate distance measures up front for these tests, and report both (at least in supplementary material), as the choice of distance measure can affect results and conclusions!</p>
</section>
<section id="covariate-adjusted-permanova" class="level3">
<h3 class="anchored" data-anchor-id="covariate-adjusted-permanova">Covariate-adjusted PERMANOVA</h3>
<p>You can also adjust for covariates in PERMANOVA, and often should, depending on your study design. Let‚Äôs fit a more complex model, adjusting for infant sex, birth weight, and the total number of assigned reads.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="fl">2.5</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_agg</span>(<span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist_calc</span>(<span class="at">dist =</span> <span class="st">"bray"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist_permanova</span>(</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">variables =</span> <span class="fu">c</span>(<span class="st">"birth_mode"</span>, <span class="st">"sex"</span>, <span class="st">"birth_weight"</span>, <span class="st">"number_reads"</span>),</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_perms =</span> <span class="dv">99</span>, <span class="at">seed =</span> <span class="dv">111</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">perm_get</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Dropping samples with missings: 15</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2023-05-23 20:39:12.478693 - Starting PERMANOVA with 99 perms with 1 processes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2023-05-23 20:39:13.167696 - Finished PERMANOVA</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for adonis under reduced model
Marginal effects of terms
Permutation: free
Number of permutations: 99

vegan::adonis2(formula = formula, data = metadata, permutations = n_perms, by = by, parallel = parall)
              Df SumOfSqs      R2       F Pr(&gt;F)   
birth_mode     1   10.794 0.10163 34.8045   0.01 **
sex            1    0.280 0.00264  0.9031   0.43   
birth_weight   1    0.565 0.00532  1.8215   0.06 . 
number_reads   1    2.873 0.02705  9.2656   0.01 **
Residual     286   88.696 0.83509                  
Total        290  106.211 1.00000                  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use more permutations for a more reliable p.value in your real work (slower)</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Set a random seed number for reproducibility of this stochastic method</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="pca" class="level2">
<h2 class="anchored" data-anchor-id="pca">PCA</h2>
<ul>
<li>Principal <strong>Components</strong> Analysis.</li>
<li>For practical purposes, PCA is quite similar to Principal Co-ordinates Analysis.</li>
<li>In fact, PCA produces equivalent results to PCoA with Euclidean distances.</li>
</ul>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
‚ÄúHelp, what are Euclidean distances?‚Äù
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>Euclidean distances are essentially a generalization of Pythagoras‚Äô theorem to more dimensions.</li>
<li>In our data every taxon is a feature, a dimension, on which we calculate Euclidean distances.</li>
</ul>
<p><strong>Pythagoras‚Äô theorem:</strong></p>
<p><span class="math display">\[c = \sqrt{a^2 + b^2}\]</span></p>
<p><strong>Euclidean distance:</strong></p>
<p><span class="math display">\[d\left(p, q\right) = \sqrt{\sum _{i=1}^{n_{taxa}} \left( p_{i}-q_{i}\right)^2 }\]</span></p>
<ul>
<li>Distance <span class="math inline">\(d\)</span> between samples <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span>, with <span class="math inline">\(i\)</span> taxa.</li>
</ul>
</div>
</div>
</div>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true" href="">Euclidean PCoA</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false" href="">PCA on counts</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_agg</span>(<span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist_calc</span>(<span class="at">dist =</span> <span class="st">"euclidean"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_calc</span>(<span class="at">method =</span> <span class="st">"PCoA"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_plot</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rug</span>(<span class="at">alpha =</span> <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="576"></p>
</div>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_agg</span>(<span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_calc</span>(<span class="at">method =</span> <span class="st">"PCA"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_plot</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rug</span>(<span class="at">alpha =</span> <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="576"></p>
</div>
</div>
</div>
</div>
</div>
<p><strong>Problems with PCA (or PCoA with Euclidean Distances) on microbiome data</strong></p>
<ul>
<li>These plots look weird! most samples bunch in the middle, with spindly projections..</li>
<li>Sensitive to sparsity (double-zero problem) ‚Äì&gt; filter rare taxa</li>
<li>Excessive emphasis on high-abundance taxa ‚Äì&gt; log transform features first</li>
</ul>
<hr>
</section>
<section id="log-transformations-and-clr" class="level2">
<h2 class="anchored" data-anchor-id="log-transformations-and-clr">Log transformations, and CLR</h2>
<ul>
<li>First let‚Äôs look at the abundance again, this time with heatmaps.</li>
<li>Each column is a sample (from an infant), and each row is a taxon.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_sort</span>(<span class="at">by =</span> sum, <span class="at">at =</span> <span class="st">"genus"</span>, <span class="at">trans =</span> <span class="st">"compositional"</span>, <span class="at">tree_warn =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="at">trans =</span> <span class="st">"compositional"</span>, <span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_heatmap</span>(<span class="at">samples =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">taxa =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">name =</span> <span class="st">"Proportions"</span>, <span class="at">tax_seriation =</span> <span class="st">"Identity"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<ul>
<li>Even though we have picked the top 20 most abundant genera, there are still a lot of zeros</li>
<li><strong>Problem:</strong> We need to deal with the zeros, because <code>log(0)</code> is undefined.</li>
<li><strong>Solution:</strong> add a small amount to every value (or just every zero), before applying the log transformation.</li>
<li>This small value is often called a pseudo-count.</li>
</ul>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
What value should we use for the pseudo-count?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>One easy option is to just add a count of 1</li>
<li>Another popular option is to add half of the smallest observed real value (from across the whole dataset)</li>
<li>In general, for zero replacement, keep it simple and <strong>record your approach</strong></li>
</ul>
</div>
</div>
</div>
<section id="centered-log-ratio-transformation" class="level3">
<h3 class="anchored" data-anchor-id="centered-log-ratio-transformation">Centered Log Ratio transformation:</h3>
<p><strong>Remember</strong>, <a href="https://doi.org/10.3389/fmicb.2017.02224">Microbiome Datasets Are Compositional: And This Is Not Optional.</a></p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
More details on the ‚ÄúCoDa‚Äù problem:
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>The sequencing data gives us relative abundances, not absolute abundances.</li>
<li>The total number of reads sequenced per sample is an arbitrary total.</li>
</ul>
<p><strong>This leads to two main types of problem:</strong></p>
<ul>
<li>Interpretation caveats: see differential abundance section later</li>
<li>Statistical issues: taxon abundances are not independent, and may appear negatively correlated</li>
<li>These issues are worse with simpler ecosystems</li>
</ul>
<p>Example: If one taxon blooms, the relative abundance of all other taxa will appear to decrease, even if they did not.</p>
<p><em>There is the same problem in theory with RNAseq data, but I suspect it is less bothersome because there are many more competing ‚Äúspecies‚Äù of RNA transcript than there are bacterial species in even a very complex microbiome.</em> <em>The centered-log-ratio transformation (along with some other similar ratio transformations) are claimed to help with the statistical issues by transforming the abundances from the simplex to the real space.</em></p>
</div>
</div>
</div>
<p><strong>TL;DR - the CLR transformation is useful for compositional microbiome data.</strong></p>
<ul>
<li>Practically, the CLR transformation involves finding the geometric mean of each sample</li>
<li>Then dividing abundance of each taxon in that sample by this geometric mean</li>
<li>Finally, you take the natural log of these ratios</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_sort</span>(<span class="at">by =</span> sum, <span class="at">at =</span> <span class="st">"genus"</span>, <span class="at">trans =</span> <span class="st">"compositional"</span>, <span class="at">tree_warn =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_agg</span>(<span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="at">trans =</span> <span class="st">"clr"</span>, <span class="at">zero_replace =</span> <span class="st">"halfmin"</span>, <span class="at">chain =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_heatmap</span>(</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">samples =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">taxa =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">grid_lwd =</span> <span class="dv">2</span>, <span class="at">name =</span> <span class="st">"CLR"</span>,</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">colors =</span> <span class="fu">heat_palette</span>(<span class="at">sym =</span> <span class="cn">TRUE</span>),</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_seriation =</span> <span class="st">"Identity"</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="576"></p>
</div>
</div>
</section>
<section id="pca-on-clr-transformed-taxa" class="level3">
<h3 class="anchored" data-anchor-id="pca-on-clr-transformed-taxa">PCA on CLR-transformed taxa</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="fl">2.5</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="at">rank =</span> <span class="st">"genus"</span>, <span class="at">trans =</span> <span class="st">"clr"</span>, <span class="at">zero_replace =</span> <span class="st">"halfmin"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_calc</span>(<span class="at">method =</span> <span class="st">"PCA"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_plot</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">"birth_mode"</span>) <span class="sc">+</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>(<span class="dv">12</span>) <span class="sc">+</span></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>(<span class="fl">0.7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<ul>
<li>After the CLR transformation, the plot looks better</li>
<li>We can see a pattern where the gut microbiomes of infants cluster by birth mode</li>
</ul>
<section id="so-why-is-pca-interesting-for-us" class="level4">
<h4 class="anchored" data-anchor-id="so-why-is-pca-interesting-for-us">So why is PCA interesting for us?</h4>
<ul>
<li><p>Principal components are built directly from a (linear) combination of the original features.</p></li>
<li><p>That means we know how much each taxon contributes to each PC dimension</p></li>
<li><p>We can plot this information (loadings) as arrows, alongside the sample points</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> infants <span class="sc">%&gt;%</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="fl">2.5</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="at">rank =</span> <span class="st">"genus"</span>, <span class="at">trans =</span> <span class="st">"clr"</span>, <span class="at">zero_replace =</span> <span class="st">"halfmin"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_calc</span>(<span class="at">method =</span> <span class="st">"PCA"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_plot</span>(</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">size =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">"birth_mode"</span>,</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">plot_taxa =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="at">tax_vec_length =</span> <span class="fl">0.275</span>,</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_lab_style =</span> <span class="fu">tax_lab_style</span>(</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">type =</span> <span class="st">"text"</span>, <span class="at">max_angle =</span> <span class="dv">90</span>, <span class="at">aspect_ratio =</span> <span class="fl">0.7</span>,</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">size =</span> <span class="dv">3</span>, <span class="at">fontface =</span> <span class="st">"bold"</span></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>(<span class="dv">12</span>) <span class="sc">+</span></span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>(<span class="fl">0.7</span>, <span class="at">clip =</span> <span class="st">"off"</span>)</span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>pca</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>Interestingly, samples on the right of the plot (which tend to be vaginally-delivered infants) seem to have relatively more Bifidobacterium, Bacteroides and Escherichia, whilst the C-section born infants have relatively more Veillonella.</p>
<div class="callout-important callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Wait, how to interpret these taxa loadings?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="cautiously" class="level5">
<h5 class="anchored" data-anchor-id="cautiously">Cautiously</h5>
<ul>
<li>There are caveats and nuance to the interpretation of these plots, which are called PCA bi-plots</li>
<li>You can read more here: <a href="https://sites.google.com/site/mb3gustame/indirect-gradient-analysis/pca" class="uri">https://sites.google.com/site/mb3gustame/indirect-gradient-analysis/pca</a></li>
</ul>
</section>
<section id="in-general" class="level5">
<h5 class="anchored" data-anchor-id="in-general">In general:</h5>
<p>The relative length and direction of an arrow indicates how much that taxon contributes to the variation on each visible PC axis, e.g.&nbsp;Variation in Enterococcus abundance contributes quite a lot to variation along the PC2 axis.</p>
<p>This allows you to infer that samples positioned at the bottom of the plot will tend to have higher relative abundance of Enterococcus than samples at the top of the plot.</p>
<p><em>(Side note, Phocaeicola were considered part of Bacteroides until this year!)</em></p>
</section>
</div>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-12-contents" aria-controls="callout-12" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Fancy circular bar charts?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-12" class="callout-12-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>We can make another kind of barplot, using the PCA information to order our samples in a circular layout.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>iris <span class="ot">&lt;-</span> infants <span class="sc">%&gt;%</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="fl">2.5</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="at">rank =</span> <span class="st">"genus"</span>, <span class="at">trans =</span> <span class="st">"clr"</span>, <span class="at">zero_replace =</span> <span class="st">"halfmin"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_calc</span>(<span class="at">method =</span> <span class="st">"PCA"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ord_plot_iris</span>(</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_level =</span> <span class="st">"genus"</span>, <span class="at">n_taxa =</span> <span class="dv">12</span>, <span class="at">other =</span> <span class="st">"Other"</span>,</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">anno_colour =</span> <span class="st">"birth_mode"</span>,</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">anno_colour_style =</span> <span class="fu">list</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">size =</span> <span class="fl">0.6</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span>)</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>patchwork<span class="sc">::</span><span class="fu">wrap_plots</span>(pca, iris, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</div>
</div>
</div>
<hr>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Taxon filtering
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>We probably want to filter out <strong>rare</strong> taxa, before performing some kinds of analysis.</p>
<section id="why-remove-rare-taxa" class="level5">
<h5 class="anchored" data-anchor-id="why-remove-rare-taxa">Why remove rare taxa?</h5>
<p><strong>Rare taxa might sometimes be:</strong></p>
<ol type="1">
<li>Sequencing errors</li>
<li>Statistically problematic</li>
<li>Biologically irrelevant</li>
</ol>
</section>
<section id="how-to-remove-rare-taxa" class="level5">
<h5 class="anchored" data-anchor-id="how-to-remove-rare-taxa">How to remove rare taxa?</h5>
<p><strong>What is rare?</strong> Two main concepts.</p>
<ul>
<li>Low <strong>prevalence</strong> - taxon only detected in a small number of samples in your dataset.</li>
<li>Low <strong>abundance</strong> - relatively few reads assigned to that taxon (on average or in total)</li>
</ul>
<p>Considering the impact of issues 1, 2, and 3, let‚Äôs say we are not interested in Species that occur in fewer than 2% of samples, and they have to have at least 10,000 reads in total across all samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ntaxa</span>(shao19) <span class="co"># before filtering</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 819</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>shao19 <span class="sc">%&gt;%</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="dv">2</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">min_total_abundance =</span> <span class="dv">10000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ntaxa</span>() <span class="co"># after filtering</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Proportional min_prevalence given: 0.02 --&gt; min 33/1644 samples.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 253</code></pre>
</div>
</div>
<ul>
<li>Wow so that would remove <strong>most</strong> of our unique taxa!</li>
<li>What is going on? Let‚Äôs make some plots!</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make table of summary statistics for the unique taxa in shao19</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>shaoTaxaStats <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">taxon =</span> <span class="fu">taxa_names</span>(shao19),</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">prevalence =</span> microbiome<span class="sc">::</span><span class="fu">prevalence</span>(shao19),</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">total_abundance =</span> <span class="fu">taxa_sums</span>(shao19)</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<details>
<summary>Some ggplot2 code</summary>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> shaoTaxaStats <span class="sc">%&gt;%</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(total_abundance, prevalence)) <span class="sc">+</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rug</span>(<span class="at">alpha =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">label_number</span>(), <span class="at">name =</span> <span class="st">"Total Abundance"</span>) <span class="sc">+</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">label_percent</span>(), <span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">breaks_pretty</span>(<span class="at">n =</span> <span class="dv">9</span>),</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">name =</span> <span class="st">"Prevalence (%)"</span>,</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">sec.axis =</span> <span class="fu">sec_axis</span>(</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">trans =</span> <span class="sc">~</span> . <span class="sc">*</span> <span class="fu">nsamples</span>(shao19), <span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">breaks_pretty</span>(<span class="at">n =</span> <span class="dv">9</span>),</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">name =</span> <span class="st">"Prevalence (N samples)"</span></span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>So most taxa have a low prevalence, and handful have way more reads than most.</p>
<p>Let‚Äôs label those points to check which taxa are the big time players.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>p <span class="sc">+</span> ggrepel<span class="sc">::</span><span class="fu">geom_text_repel</span>(</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> <span class="cf">function</span>(df) <span class="fu">filter</span>(df, total_abundance <span class="sc">&gt;</span> <span class="fl">1e9</span> <span class="sc">|</span> prevalence <span class="sc">&gt;</span> <span class="fl">0.6</span>),</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">label =</span> taxon), <span class="at">size =</span> <span class="fl">2.5</span>, <span class="at">min.segment.length =</span> <span class="dv">0</span>, <span class="at">force =</span> <span class="dv">15</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<p>Those taxa make sense for this dataset of mostly infant gut microbiome samples.</p>
<p>Now let‚Äôs zoom in on the less abundant taxa by log-transforming the axes. We‚Äôll also add lines indicating the thresholds of 2% prevalence and 10000 reads abundance.</p>
<div class="cell">
<details>
<summary>Some more ggplot2 code</summary>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>shaoTaxaStats <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> total_abundance, <span class="at">y =</span> prevalence)) <span class="sc">+</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">10000</span>, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="st">"dotted"</span>) <span class="sc">+</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">2</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="st">"dotted"</span>) <span class="sc">+</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rug</span>(<span class="at">alpha =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>(<span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">label_number</span>(), <span class="at">name =</span> <span class="st">"Total Abundance"</span>) <span class="sc">+</span></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>(</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">label_percent</span>(), <span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">breaks_log</span>(<span class="at">n =</span> <span class="dv">9</span>),</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">name =</span> <span class="st">"Prevalence (%)"</span>,</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">sec.axis =</span> <span class="fu">sec_axis</span>(</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>      <span class="at">trans =</span> <span class="sc">~</span> . <span class="sc">*</span> <span class="fu">nsamples</span>(shao19), <span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">breaks_log</span>(<span class="at">n =</span> <span class="dv">9</span>),</span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">name =</span> <span class="st">"Prevalence (N samples)"</span></span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<ul>
<li>We can break this down by phylum if we add the taxonomic table information.</li>
</ul>
<div class="cell">
<details>
<summary>A lot more ggplot2 code!</summary>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># don't worry about this code if it's confusing, just focus on the plot output</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>shao19 <span class="sc">%&gt;%</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_table</span>() <span class="sc">%&gt;%</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>(<span class="at">rownames =</span> <span class="st">"taxon"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(shaoTaxaStats, <span class="at">by =</span> <span class="st">"taxon"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_count</span>(phylum, <span class="at">name =</span> <span class="st">"phylum_count"</span>, <span class="at">sort =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">phylum =</span> <span class="fu">factor</span>(phylum, <span class="at">levels =</span> <span class="fu">unique</span>(phylum))) <span class="sc">%&gt;%</span> <span class="co"># to fix facet order</span></span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">phylum =</span> forcats<span class="sc">::</span><span class="fu">fct_lump_n</span>(phylum, <span class="at">n =</span> <span class="dv">5</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">phylum =</span> forcats<span class="sc">::</span><span class="fu">fct_explicit_na</span>(phylum, <span class="at">na_level =</span> <span class="st">"Other"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(total_abundance, prevalence)) <span class="sc">+</span></span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">10000</span>, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="st">"dotted"</span>) <span class="sc">+</span></span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">2</span> <span class="sc">/</span> <span class="dv">100</span>, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="st">"dotted"</span>) <span class="sc">+</span></span>
<span id="cb50-14"><a href="#cb50-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb50-15"><a href="#cb50-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rug</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb50-16"><a href="#cb50-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>(</span>
<span id="cb50-17"><a href="#cb50-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">label_log</span>(), <span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">breaks_log</span>(<span class="at">n =</span> <span class="dv">5</span>),</span>
<span id="cb50-18"><a href="#cb50-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">name =</span> <span class="st">"Total Abundance"</span></span>
<span id="cb50-19"><a href="#cb50-19" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb50-20"><a href="#cb50-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>(</span>
<span id="cb50-21"><a href="#cb50-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> scales<span class="sc">::</span><span class="fu">label_percent</span>(), <span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">breaks_log</span>(<span class="at">n =</span> <span class="dv">9</span>),</span>
<span id="cb50-22"><a href="#cb50-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">name =</span> <span class="st">"Prevalence (%)"</span>,</span>
<span id="cb50-23"><a href="#cb50-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">sec.axis =</span> <span class="fu">sec_axis</span>(</span>
<span id="cb50-24"><a href="#cb50-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">trans =</span> <span class="sc">~</span> . <span class="sc">*</span> <span class="fu">nsamples</span>(shao19), <span class="at">breaks =</span> scales<span class="sc">::</span><span class="fu">breaks_log</span>(<span class="at">n =</span> <span class="dv">9</span>),</span>
<span id="cb50-25"><a href="#cb50-25" aria-hidden="true" tabindex="-1"></a>      <span class="at">name =</span> <span class="st">"Prevalence (N samples)"</span></span>
<span id="cb50-26"><a href="#cb50-26" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb50-27"><a href="#cb50-27" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb50-28"><a href="#cb50-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="st">"phylum"</span>) <span class="sc">+</span></span>
<span id="cb50-29"><a href="#cb50-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: There was 1 warning in `mutate()`.
‚Ñπ In argument: `phylum = forcats::fct_explicit_na(phylum, na_level = "Other")`.
Caused by warning:
! `fct_explicit_na()` was deprecated in forcats 1.0.0.
‚Ñπ Please use `fct_na_value_to_level()` instead.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p><strong>How you pick a threshold, depends on what analysis method you are filtering for!</strong></p>
<ul>
<li>alpha diversity: do not filter</li>
<li>beta diversity: relevance of threshold depends on your distance measure</li>
<li>differential abundance testing: stringent filtering, prevalence &gt;5%, &gt;10%?</li>
</ul>
</section>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="taxon-stats" class="level2">
<h2 class="anchored" data-anchor-id="taxon-stats">Taxon stats</h2>
<p>From the PCA loadings and barplots above, we have some strong suspicions about which taxa have a higher relative abundance in vaginally delivered infants than in c-section delivered infants, and vice versa, but we can also statistically test this. This is often called ‚Äúdifferential abundance‚Äù (DA) testing, in the style of ‚Äúdifferential expression‚Äù (DE) testing from the transcriptomics field.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">comp_barplot</span>(</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">tax_level =</span> <span class="st">"genus"</span>, <span class="at">n_taxa =</span> <span class="dv">12</span>, <span class="at">facet_by =</span> <span class="st">"birth_mode"</span>,</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">label =</span> <span class="cn">NULL</span>, <span class="at">bar_outline_colour =</span> <span class="cn">NA</span></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<section id="model-one-taxon" class="level3">
<h3 class="anchored" data-anchor-id="model-one-taxon">Model one taxon</h3>
<ul>
<li>We will start by creating a linear regression model for one genus, Bacteroides.</li>
<li>We will transform the count data by first making it proportions, and then taking a base 2 logarithm, log2, after adding a pseudocount.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>bacteroidesRegression1 <span class="ot">&lt;-</span> infants <span class="sc">%&gt;%</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="st">"compositional"</span>, <span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="st">"log2"</span>, <span class="at">zero_replace =</span> <span class="st">"halfmin"</span>, <span class="at">chain =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_model</span>(</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">type =</span> <span class="st">"lm"</span>, <span class="at">rank =</span> <span class="st">"genus"</span>,</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">taxa =</span> <span class="st">"Bacteroides"</span>, <span class="at">variables =</span> <span class="st">"birth_mode"</span>,</span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">return_psx =</span> <span class="cn">FALSE</span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pluck</span>(<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Modelling: Bacteroides</code></pre>
</div>
</div>
<ul>
<li>Looking at the regression results</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bacteroidesRegression1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
Bacteroides ~ birth_mode

Residuals:
    Min      1Q  Median      3Q     Max 
-7.7492 -0.6172 -0.6172  2.6421 18.0804 

Coefficients:
                  Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)       -19.3756     0.4863  -39.84   &lt;2e-16 ***
birth_modevaginal   7.1320     0.6812   10.47   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 5.957 on 304 degrees of freedom
Multiple R-squared:  0.265, Adjusted R-squared:  0.2626 
F-statistic: 109.6 on 1 and 304 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>broom<span class="sc">::</span><span class="fu">tidy</span>(bacteroidesRegression1, <span class="at">conf.int =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 √ó 7
  term              estimate std.error statistic   p.value conf.low conf.high
  &lt;chr&gt;                &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;
1 (Intercept)         -19.4      0.486     -39.8 1.08e-122   -20.3     -18.4 
2 birth_modevaginal     7.13     0.681      10.5 4.13e- 22     5.79      8.47</code></pre>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-14-contents" aria-controls="callout-14" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Click here for optional ggplot2 extension exercise:
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-14" class="callout-14-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Starting from a dataframe like the one produced by the code below, plot:</p>
<ol type="1">
<li>Easy: The percentage prevalence of Bacteroides in each birth_mode group</li>
<li>Medium: The distribution of relative abundance of Bacteroides in each birth_mode group, omitting zeros, on a log2 scale</li>
<li>Hard: Do task 1 or 2 for for several taxa in one plot - (hint: <code>pivot_longer</code>)</li>
</ol>
<div class="sourceCode" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>infants <span class="sc">%&gt;%</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="st">"compositional"</span>, <span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_get</span>() <span class="sc">%&gt;%</span></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_otu2samdat</span>(<span class="at">taxa =</span> <span class="st">"Bacteroides"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">samdat_tbl</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>We can fit a model with covariates, as we did for PERMANOVA</p>
<ul>
<li>We will convert the categorical variables into indicator (dummy) variables</li>
<li>We will scale the continuous covariates to 0 mean and SD 1 (z-scores)</li>
<li>You‚Äôll see this will make our subsequent plots easier to interpret later</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>infants <span class="ot">&lt;-</span> infants <span class="sc">%&gt;%</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ps_mutate</span>(</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">C_section =</span> <span class="fu">if_else</span>(birth_mode <span class="sc">==</span> <span class="st">"c_section"</span>, <span class="at">true =</span> <span class="dv">1</span>, <span class="at">false =</span> <span class="dv">0</span>),</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">Female =</span> <span class="fu">if_else</span>(sex <span class="sc">==</span> <span class="st">"female"</span>, <span class="at">true =</span> <span class="dv">1</span>, <span class="at">false =</span> <span class="dv">0</span>),</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">Birth_weight_Z =</span> <span class="fu">scale</span>(birth_weight, <span class="at">center =</span> <span class="cn">TRUE</span>, <span class="at">scale =</span> <span class="cn">TRUE</span>),</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">Reads_Z =</span> <span class="fu">scale</span>(number_reads, <span class="at">center =</span> <span class="cn">TRUE</span>, <span class="at">scale =</span> <span class="cn">TRUE</span>)</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>bacteroidesRegression2 <span class="ot">&lt;-</span> infants <span class="sc">%&gt;%</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="st">"compositional"</span>, <span class="at">rank =</span> <span class="st">"genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="st">"log2"</span>, <span class="at">zero_replace =</span> <span class="st">"halfmin"</span>, <span class="at">chain =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_model</span>(</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">type =</span> <span class="st">"lm"</span>, <span class="at">rank =</span> <span class="st">"genus"</span>, <span class="at">taxa =</span> <span class="st">"Bacteroides"</span>,</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">variables =</span> <span class="fu">c</span>(<span class="st">"C_section"</span>, <span class="st">"Female"</span>, <span class="st">"Birth_weight_Z"</span>, <span class="st">"Reads_Z"</span>),</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">return_psx =</span> <span class="cn">FALSE</span></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pluck</span>(<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in do.call(fun, list(txt)): 15 / 306 values are NA in Female</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in do.call(fun, list(txt)): 14 / 306 values are NA in Birth_weight_Z</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Modelling: Bacteroides</code></pre>
</div>
</div>
<ul>
<li>Looking at the regression results</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(bacteroidesRegression2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
Bacteroides ~ C_section + Female + Birth_weight_Z + Reads_Z

Residuals:
    Min      1Q  Median      3Q     Max 
-9.4271 -2.1555 -0.4115  2.8176 18.1784 

Coefficients:
               Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)    -11.7942     0.6103 -19.325   &lt;2e-16 ***
C_section       -7.5696     0.7206 -10.505   &lt;2e-16 ***
Female          -0.3809     0.7101  -0.536    0.592    
Birth_weight_Z   0.3277     0.3514   0.932    0.352    
Reads_Z          0.5361     0.3620   1.481    0.140    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 5.934 on 286 degrees of freedom
  (15 observations deleted due to missingness)
Multiple R-squared:  0.2854,    Adjusted R-squared:  0.2754 
F-statistic: 28.55 on 4 and 286 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>broom<span class="sc">::</span><span class="fu">tidy</span>(bacteroidesRegression2, <span class="at">conf.int =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 √ó 7
  term           estimate std.error statistic  p.value conf.low conf.high
  &lt;chr&gt;             &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;
1 (Intercept)     -11.8       0.610   -19.3   8.15e-54  -13.0      -10.6 
2 C_section        -7.57      0.721   -10.5   4.81e-22   -8.99      -6.15
3 Female           -0.381     0.710    -0.536 5.92e- 1   -1.78       1.02
4 Birth_weight_Z    0.328     0.351     0.932 3.52e- 1   -0.364      1.02
5 Reads_Z           0.536     0.362     1.48  1.40e- 1   -0.176      1.25</code></pre>
</div>
</div>
<section id="many-da-methods" class="level4">
<h4 class="anchored" data-anchor-id="many-da-methods">Many DA methods</h4>
<ul>
<li>This method simple method is borrowed from MaAsLin2</li>
<li>Note: they call the compositional transformation ‚ÄúTotal Sum Scaling (TSS)‚Äù)</li>
<li>This is quite a straightforward method, so we will stick with this for today</li>
<li>But, many statistical methods have been developed for differential abundance analyses</li>
</ul>
<p>Microbiome abundance data are quite awkward, statistically speaking, due to their sparseness and compositionality. Each successive method claims to handle some aspect of this awkwardness ‚Äúbetter‚Äù than previous methods.</p>
<p>The aim is to have a method with adequate power to detect true associations, whilst controlling the type 1 error rate, the ‚Äúfalse positive‚Äù reporting of associations that are not ‚Äútruly‚Äù present.</p>
<p>Results are surprisingly inconsistent across the different methods, as demonstrated this year in a <a href="https://www.nature.com/articles/s41467-022-28034-z">fascinating analysis by Jacob Nearing and colleagues</a>.</p>
</section>
<section id="so-what-to-do" class="level4">
<h4 class="anchored" data-anchor-id="so-what-to-do">So, what to do?</h4>
<ul>
<li>Filter out the noise &amp; interpret results with caution! use multiple testing corrections</li>
<li>Remember it‚Äôs all relative (abundance)</li>
<li>Try 2 or 3 methods and/or use same method as a previous study if replicating
<ul>
<li>Avoid Lefse and edgeR?</li>
<li>Beware: Not all methods allow covariate adjustment &amp; few allow random effects (for time-series)</li>
</ul></li>
</ul>
</section>
</section>
<section id="model-all-the-taxa" class="level3">
<h3 class="anchored" data-anchor-id="model-all-the-taxa">Model all the taxa!</h3>
<ol type="1">
<li>We‚Äôre not normally interested in just one taxon!</li>
<li>It‚Äôs also hard to decide which taxonomic rank we are most interested in modelling!
<ul>
<li>Lower ranks like species or ASVs give better resolution but also more sparsity and classification uncertainty‚Ä¶</li>
<li>Higher ranks e.g.&nbsp;classes, could also be more powerful if you think most taxa within that class will follow a similar pattern.</li>
</ul></li>
</ol>
<ul>
<li>So now we will fit a similar model for almost every taxon* at every rank from phylum to species</li>
<li>*We‚Äôll filter out species with a prevalence of less than 10%</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The code for `taxatree_models` is quite similar to tax_model.</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="co"># However, you might need to run `tax_prepend_ranks` to ensure that each taxon at each rank is always unique.</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>shaoModels <span class="ot">&lt;-</span> infants <span class="sc">%&gt;%</span></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_prepend_ranks</span>() <span class="sc">%&gt;%</span></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="st">"compositional"</span>, <span class="at">rank =</span> <span class="st">"species"</span>, <span class="at">keep_counts =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_filter</span>(<span class="at">min_prevalence =</span> <span class="fl">0.1</span>, <span class="at">undetected =</span> <span class="dv">0</span>, <span class="at">use_counts =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_transform</span>(<span class="at">trans =</span> <span class="st">"log2"</span>, <span class="at">chain =</span> <span class="cn">TRUE</span>, <span class="at">zero_replace =</span> <span class="st">"halfmin"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">taxatree_models</span>(</span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">type =</span> lm,</span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">ranks =</span> <span class="fu">c</span>(<span class="st">"phylum"</span>, <span class="st">"class"</span>, <span class="st">"order"</span>, <span class="st">"family"</span>, <span class="st">"genus"</span>, <span class="st">"species"</span>),</span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">variables =</span> <span class="fu">c</span>(<span class="st">"C_section"</span>, <span class="st">"Female"</span>, <span class="st">"Birth_weight_Z"</span>, <span class="st">"Reads_Z"</span>)</span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Proportional min_prevalence given: 0.1 --&gt; min 31/306 samples.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2023-05-23 20:39:23.285753 - modelling at rank: phylum</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in do.call(fun, list(txt)): 15 / 306 values are NA in Female</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in do.call(fun, list(txt)): 14 / 306 values are NA in Birth_weight_Z</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2023-05-23 20:39:23.34296 - modelling at rank: class</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in do.call(fun, list(txt)): 15 / 306 values are NA in Female

Warning in do.call(fun, list(txt)): 14 / 306 values are NA in Birth_weight_Z</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2023-05-23 20:39:23.428134 - modelling at rank: order</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in do.call(fun, list(txt)): 15 / 306 values are NA in Female

Warning in do.call(fun, list(txt)): 14 / 306 values are NA in Birth_weight_Z</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2023-05-23 20:39:23.526922 - modelling at rank: family</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in do.call(fun, list(txt)): 15 / 306 values are NA in Female

Warning in do.call(fun, list(txt)): 14 / 306 values are NA in Birth_weight_Z</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2023-05-23 20:39:23.655146 - modelling at rank: genus</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in do.call(fun, list(txt)): 15 / 306 values are NA in Female

Warning in do.call(fun, list(txt)): 14 / 306 values are NA in Birth_weight_Z</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2023-05-23 20:39:23.796617 - modelling at rank: species</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in do.call(fun, list(txt)): 15 / 306 values are NA in Female

Warning in do.call(fun, list(txt)): 14 / 306 values are NA in Birth_weight_Z</code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>shaoModels</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>psExtra object - a phyloseq object with extra slots:

phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 39 taxa and 306 samples ]
sample_data() Sample Data:       [ 306 samples by 15 sample variables ]
tax_table()   Taxonomy Table:    [ 39 taxa by 6 taxonomic ranks ]

otu_get(counts = TRUE)       [ 39 taxa and 306 samples ]

psExtra info:
tax_agg = "species" tax_trans = "compositional&amp;log2" 

taxatree_models list:
Ranks: phylum/class/order/family/genus/species</code></pre>
</div>
</div>
<p><em>Why filter the taxa? It‚Äôs less likely that we are interested in rare taxa, and models of rare taxon abundances are more likely to be unreliable. Reducing the the number of taxa modelled also makes the process faster and makes visualizing the results easier!</em></p>
<section id="getting-stats-from-the-models" class="level4">
<h4 class="anchored" data-anchor-id="getting-stats-from-the-models">Getting stats from the models</h4>
<p>Next we will get a data.frame containing the regression coefficient estimates, test statistics and corresponding p values from all these regression models.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>shaoStats <span class="ot">&lt;-</span> <span class="fu">taxatree_models2stats</span>(shaoModels)</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>shaoStats</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>psExtra object - a phyloseq object with extra slots:

phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 39 taxa and 306 samples ]
sample_data() Sample Data:       [ 306 samples by 15 sample variables ]
tax_table()   Taxonomy Table:    [ 39 taxa by 6 taxonomic ranks ]

otu_get(counts = TRUE)       [ 39 taxa and 306 samples ]

psExtra info:
tax_agg = "species" tax_trans = "compositional&amp;log2" 

taxatree_stats dataframe:
89 taxa at 6 ranks: phylum, class, order, family, genus, species 
4 terms: C_section, Female, Birth_weight_Z, Reads_Z</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>shaoStats <span class="sc">%&gt;%</span> <span class="fu">taxatree_stats_get</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 356 √ó 8
   term           taxon      rank  formula estimate std.error statistic  p.value
   &lt;fct&gt;          &lt;chr&gt;      &lt;fct&gt; &lt;chr&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
 1 C_section      p: Proteo‚Ä¶ phyl‚Ä¶ `p: Pr‚Ä¶  -3.44       1.31    -2.62   9.23e- 3
 2 Female         p: Proteo‚Ä¶ phyl‚Ä¶ `p: Pr‚Ä¶   0.595      1.29     0.460  6.46e- 1
 3 Birth_weight_Z p: Proteo‚Ä¶ phyl‚Ä¶ `p: Pr‚Ä¶   0.0179     0.640    0.0279 9.78e- 1
 4 Reads_Z        p: Proteo‚Ä¶ phyl‚Ä¶ `p: Pr‚Ä¶  -1.22       0.659   -1.86   6.44e- 2
 5 C_section      p: Actino‚Ä¶ phyl‚Ä¶ `p: Ac‚Ä¶ -16.9        2.37    -7.12   8.98e-12
 6 Female         p: Actino‚Ä¶ phyl‚Ä¶ `p: Ac‚Ä¶  -2.80       2.33    -1.20   2.30e- 1
 7 Birth_weight_Z p: Actino‚Ä¶ phyl‚Ä¶ `p: Ac‚Ä¶   1.65       1.15     1.43   1.55e- 1
 8 Reads_Z        p: Actino‚Ä¶ phyl‚Ä¶ `p: Ac‚Ä¶  -2.86       1.19    -2.41   1.67e- 2
 9 C_section      p: Firmic‚Ä¶ phyl‚Ä¶ `p: Fi‚Ä¶  15.3        4.21     3.62   3.43e- 4
10 Female         p: Firmic‚Ä¶ phyl‚Ä¶ `p: Fi‚Ä¶   0.741      4.15     0.179  8.58e- 1
# ‚Ñπ 346 more rows</code></pre>
</div>
</div>
</section>
<section id="adjusting-p-values" class="level4">
<h4 class="anchored" data-anchor-id="adjusting-p-values">Adjusting p values</h4>
<ul>
<li><p>We have performed a lot of statistical tests here!</p></li>
<li><p>It is likely that we could find some significant p-values by chance alone.</p></li>
<li><p>We should correct for multiple testing / control the false discovery rate or family-wise error rate.</p></li>
</ul>
<p><em>Instead of applying these adjustment methods across all 86 taxa models at all ranks, the default behaviour is to control the family-wise error rate per taxonomic rank.</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>shaoStats <span class="ot">&lt;-</span> shaoStats <span class="sc">%&gt;%</span> <span class="fu">taxatree_stats_p_adjust</span>(<span class="at">method =</span> <span class="st">"BH"</span>, <span class="at">grouping =</span> <span class="st">"rank"</span>)</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="co"># notice the new variable</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>shaoStats <span class="sc">%&gt;%</span> <span class="fu">taxatree_stats_get</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 356 √ó 9
# Groups:   rank [6]
   term  taxon rank  formula estimate std.error statistic  p.value p.adj.BH.rank
   &lt;fct&gt; &lt;chr&gt; &lt;fct&gt; &lt;chr&gt;      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;         &lt;dbl&gt;
 1 C_se‚Ä¶ p: P‚Ä¶ phyl‚Ä¶ `p: Pr‚Ä¶  -3.44       1.31    -2.62   9.23e- 3      2.95e- 2
 2 Fema‚Ä¶ p: P‚Ä¶ phyl‚Ä¶ `p: Pr‚Ä¶   0.595      1.29     0.460  6.46e- 1      7.38e- 1
 3 Birt‚Ä¶ p: P‚Ä¶ phyl‚Ä¶ `p: Pr‚Ä¶   0.0179     0.640    0.0279 9.78e- 1      9.78e- 1
 4 Read‚Ä¶ p: P‚Ä¶ phyl‚Ä¶ `p: Pr‚Ä¶  -1.22       0.659   -1.86   6.44e- 2      1.47e- 1
 5 C_se‚Ä¶ p: A‚Ä¶ phyl‚Ä¶ `p: Ac‚Ä¶ -16.9        2.37    -7.12   8.98e-12      7.19e-11
 6 Fema‚Ä¶ p: A‚Ä¶ phyl‚Ä¶ `p: Ac‚Ä¶  -2.80       2.33    -1.20   2.30e- 1      3.07e- 1
 7 Birt‚Ä¶ p: A‚Ä¶ phyl‚Ä¶ `p: Ac‚Ä¶   1.65       1.15     1.43   1.55e- 1      2.48e- 1
 8 Read‚Ä¶ p: A‚Ä¶ phyl‚Ä¶ `p: Ac‚Ä¶  -2.86       1.19    -2.41   1.67e- 2      4.46e- 2
 9 C_se‚Ä¶ p: F‚Ä¶ phyl‚Ä¶ `p: Fi‚Ä¶  15.3        4.21     3.62   3.43e- 4      1.83e- 3
10 Fema‚Ä¶ p: F‚Ä¶ phyl‚Ä¶ `p: Fi‚Ä¶   0.741      4.15     0.179  8.58e- 1      9.16e- 1
# ‚Ñπ 346 more rows</code></pre>
</div>
</div>
</section>
</section>
<section id="plot-all-the-taxatree_stats" class="level3">
<h3 class="anchored" data-anchor-id="plot-all-the-taxatree_stats">Plot all the taxatree_stats!</h3>
<ul>
<li><code>taxatree_plots()</code> allows you to plot statistics from all of the taxa models onto a tree layout (e.g.&nbsp;point estimates and significance).</li>
<li>The taxon model results are organised by rank, radiating out from the central root node</li>
<li>e.g.&nbsp;from Phyla around the center to Species in the outermost ring.</li>
</ul>
<p><code>taxatree_plots()</code> itself returns a list of plots, which you can arrange into one figure with the <a href="https://patchwork.data-imaginist.com/"><code>patchwork</code></a> package for example (and/or <a href="https://wilkelab.org/cowplot/articles/plot_grid.html"><code>cowplot</code></a>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>shaoStats <span class="sc">%&gt;%</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">taxatree_plots</span>(<span class="at">node_size_range =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>), <span class="at">sig_stat =</span> <span class="st">"p.adj.BH.rank"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">wrap_plots</span>(<span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid" width="576"></p>
</div>
</div>
<section id="taxatree-key" class="level4">
<h4 class="anchored" data-anchor-id="taxatree-key">Taxatree Key</h4>
<p>But how do we know which taxa are which nodes? We can create a labelled grey tree with <code>taxatree_plotkey()</code>. This labels only some of the taxa based on certain conditions that we specify.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>) <span class="co"># label position</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>key <span class="ot">&lt;-</span> shaoStats <span class="sc">%&gt;%</span></span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">taxatree_plotkey</span>(</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">taxon_renamer =</span> <span class="cf">function</span>(x) stringr<span class="sc">::</span><span class="fu">str_remove</span>(x, <span class="st">"[pfg]: "</span>),</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># conditions below, for filtering taxa to be labelled</span></span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a>    rank <span class="sc">==</span> <span class="st">"phylum"</span> <span class="sc">|</span> rank <span class="sc">==</span> <span class="st">"genus"</span> <span class="sc">&amp;</span> prevalence <span class="sc">&gt;</span> <span class="fl">0.2</span></span>
<span id="cb93-7"><a href="#cb93-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># all phyla are labelled, and all genera with a prevalence of over 0.2</span></span>
<span id="cb93-8"><a href="#cb93-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb93-9"><a href="#cb93-9" aria-hidden="true" tabindex="-1"></a>key</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="exercises_2_files/figure-html/unnamed-chunk-43-1.png" class="img-fluid" width="432"></p>
</div>
</div>
<p>You can do more with these trees to customise them to your liking. See an extended tutorial <a href="https://david-barnett.github.io/microViz/articles/web-only/modelling-taxa.html#plot-all-the-taxatree_stats">here on the microViz website</a>: including how to directly label taxa on the colored plots, change the layout and style of the trees, and even how to use a different regression modelling approach.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="co"># try it out!</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
</section>
<section id="session-info" class="level2">
<h2 class="anchored" data-anchor-id="session-info">Session info</h2>
<p><code>session_info</code> records your package versions etc. This is useful for debugging / reproducing analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">session_info</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>‚îÄ Session info ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ
 setting  value
 version  R version 4.3.0 (2023-04-21)
 os       macOS Ventura 13.3.1
 system   aarch64, darwin20
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Europe/Prague
 date     2023-05-23
 pandoc   3.0.1 @ /opt/homebrew/bin/ (via rmarkdown)

‚îÄ Packages ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ
 package          * version   date (UTC) lib source
 ade4               1.7-22    2023-02-06 [1] CRAN (R 4.3.0)
 ape                5.7-1     2023-03-13 [1] CRAN (R 4.3.0)
 backports          1.4.1     2021-12-13 [1] CRAN (R 4.3.0)
 Biobase            2.60.0    2023-04-25 [1] RSPM (R 4.3.0)
 BiocGenerics       0.46.0    2023-04-25 [1] RSPM (R 4.3.0)
 biomformat         1.28.0    2023-04-25 [1] RSPM (R 4.3.0)
 Biostrings         2.68.0    2023-04-25 [1] RSPM (R 4.3.0)
 bitops             1.0-7     2021-04-24 [1] CRAN (R 4.3.0)
 broom              1.0.4     2023-03-11 [1] CRAN (R 4.3.0)
 ca                 0.71.1    2020-01-24 [1] CRAN (R 4.3.0)
 cachem             1.0.8     2023-05-01 [1] CRAN (R 4.3.0)
 Cairo              1.6-0     2022-07-05 [1] CRAN (R 4.3.0)
 callr              3.7.3     2022-11-02 [1] CRAN (R 4.3.0)
 circlize           0.4.15    2022-05-10 [1] CRAN (R 4.3.0)
 cli                3.6.1     2023-03-23 [1] CRAN (R 4.3.0)
 clue               0.3-64    2023-01-31 [1] CRAN (R 4.3.0)
 cluster            2.1.4     2022-08-22 [1] CRAN (R 4.3.0)
 codetools          0.2-19    2023-02-01 [1] CRAN (R 4.3.0)
 colorspace         2.1-0     2023-01-23 [1] CRAN (R 4.3.0)
 ComplexHeatmap     2.16.0    2023-04-25 [1] RSPM (R 4.3.0)
 corncob            0.3.1     2022-12-05 [1] CRAN (R 4.3.0)
 crayon             1.5.2     2022-09-29 [1] CRAN (R 4.3.0)
 data.table         1.14.8    2023-02-17 [1] CRAN (R 4.3.0)
 devtools           2.4.5     2022-10-11 [1] CRAN (R 4.3.0)
 digest             0.6.31    2022-12-11 [1] CRAN (R 4.3.0)
 doParallel         1.0.17    2022-02-07 [1] CRAN (R 4.3.0)
 dplyr            * 1.1.2     2023-04-20 [1] CRAN (R 4.3.0)
 ellipsis           0.3.2     2021-04-29 [1] CRAN (R 4.3.0)
 evaluate           0.21      2023-05-05 [1] CRAN (R 4.3.0)
 fansi              1.0.4     2023-01-22 [1] CRAN (R 4.3.0)
 farver             2.1.1     2022-07-06 [1] CRAN (R 4.3.0)
 fastmap            1.1.1     2023-02-24 [1] CRAN (R 4.3.0)
 forcats          * 1.0.0     2023-01-29 [1] CRAN (R 4.3.0)
 foreach            1.5.2     2022-02-02 [1] CRAN (R 4.3.0)
 fs                 1.6.2     2023-04-25 [1] CRAN (R 4.3.0)
 generics           0.1.3     2022-07-05 [1] CRAN (R 4.3.0)
 GenomeInfoDb       1.36.0    2023-04-25 [1] RSPM (R 4.3.0)
 GenomeInfoDbData   1.2.10    2023-05-12 [1] RSPM (R 4.3.0)
 GetoptLong         1.0.5     2020-12-15 [1] CRAN (R 4.3.0)
 ggforce            0.4.1     2022-10-04 [1] CRAN (R 4.3.0)
 ggplot2          * 3.4.2     2023-04-03 [1] CRAN (R 4.3.0)
 ggraph             2.1.0     2022-10-09 [1] CRAN (R 4.3.0)
 ggrepel            0.9.3     2023-02-03 [1] CRAN (R 4.3.0)
 GlobalOptions      0.1.2     2020-06-10 [1] CRAN (R 4.3.0)
 glue               1.6.2     2022-02-24 [1] CRAN (R 4.3.0)
 graphlayouts       1.0.0     2023-05-01 [1] CRAN (R 4.3.0)
 gridExtra          2.3       2017-09-09 [1] CRAN (R 4.3.0)
 gtable             0.3.3     2023-03-21 [1] CRAN (R 4.3.0)
 hms                1.1.3     2023-03-21 [1] CRAN (R 4.3.0)
 htmltools          0.5.5     2023-03-23 [1] CRAN (R 4.3.0)
 htmlwidgets        1.6.2     2023-03-17 [1] CRAN (R 4.3.0)
 httpuv             1.6.9     2023-02-14 [1] CRAN (R 4.3.0)
 igraph             1.4.2     2023-04-07 [1] CRAN (R 4.3.0)
 IRanges            2.34.0    2023-04-25 [1] RSPM (R 4.3.0)
 iterators          1.0.14    2022-02-05 [1] CRAN (R 4.3.0)
 jsonlite           1.8.4     2022-12-06 [1] CRAN (R 4.3.0)
 knitr              1.42      2023-01-25 [1] CRAN (R 4.3.0)
 labeling           0.4.2     2020-10-20 [1] CRAN (R 4.3.0)
 later              1.3.1     2023-05-02 [1] CRAN (R 4.3.0)
 lattice            0.21-8    2023-04-05 [1] CRAN (R 4.3.0)
 lifecycle          1.0.3     2022-10-07 [1] CRAN (R 4.3.0)
 lubridate        * 1.9.2     2023-02-10 [1] CRAN (R 4.3.0)
 magrittr           2.0.3     2022-03-30 [1] CRAN (R 4.3.0)
 MASS               7.3-60    2023-05-04 [1] CRAN (R 4.3.0)
 Matrix             1.5-4     2023-04-04 [1] CRAN (R 4.3.0)
 matrixStats        0.63.0    2022-11-18 [1] CRAN (R 4.3.0)
 memoise            2.0.1     2021-11-26 [1] CRAN (R 4.3.0)
 mgcv               1.8-42    2023-03-02 [1] CRAN (R 4.3.0)
 microbiome         1.22.0    2023-04-25 [1] RSPM (R 4.3.0)
 microViz         * 0.10.10   2023-05-19 [1] Github (david-barnett/microViz@2a021f2)
 mime               0.12      2021-09-28 [1] CRAN (R 4.3.0)
 miniUI             0.1.1.1   2018-05-18 [1] CRAN (R 4.3.0)
 multtest           2.56.0    2023-04-25 [1] RSPM (R 4.3.0)
 munsell            0.5.0     2018-06-12 [1] CRAN (R 4.3.0)
 nlme               3.1-162   2023-01-31 [1] CRAN (R 4.3.0)
 patchwork          1.1.2     2022-08-19 [1] CRAN (R 4.3.0)
 permute            0.9-7     2022-01-27 [1] CRAN (R 4.3.0)
 phyloseq         * 1.44.0    2023-04-25 [1] RSPM (R 4.3.0)
 pillar             1.9.0     2023-03-22 [1] CRAN (R 4.3.0)
 pkgbuild           1.4.0     2022-11-27 [1] CRAN (R 4.3.0)
 pkgconfig          2.0.3     2019-09-22 [1] CRAN (R 4.3.0)
 pkgload            1.3.2     2022-11-16 [1] CRAN (R 4.3.0)
 plyr               1.8.8     2022-11-11 [1] CRAN (R 4.3.0)
 png                0.1-8     2022-11-29 [1] CRAN (R 4.3.0)
 polyclip           1.10-4    2022-10-20 [1] CRAN (R 4.3.0)
 prettyunits        1.1.1     2020-01-24 [1] CRAN (R 4.3.0)
 processx           3.8.1     2023-04-18 [1] CRAN (R 4.3.0)
 profvis            0.3.8     2023-05-02 [1] CRAN (R 4.3.0)
 promises           1.2.0.1   2021-02-11 [1] CRAN (R 4.3.0)
 ps                 1.7.5     2023-04-18 [1] CRAN (R 4.3.0)
 purrr            * 1.0.1     2023-01-10 [1] CRAN (R 4.3.0)
 R6                 2.5.1     2021-08-19 [1] CRAN (R 4.3.0)
 RColorBrewer       1.1-3     2022-04-03 [1] CRAN (R 4.3.0)
 Rcpp               1.0.10    2023-01-22 [1] CRAN (R 4.3.0)
 RCurl              1.98-1.12 2023-03-27 [1] CRAN (R 4.3.0)
 readr            * 2.1.4     2023-02-10 [1] CRAN (R 4.3.0)
 registry           0.5-1     2019-03-05 [1] CRAN (R 4.3.0)
 remotes            2.4.2     2021-11-30 [1] CRAN (R 4.3.0)
 reshape2           1.4.4     2020-04-09 [1] CRAN (R 4.3.0)
 rhdf5              2.44.0    2023-04-25 [1] RSPM (R 4.3.0)
 rhdf5filters       1.12.1    2023-04-30 [1] RSPM (R 4.3.0)
 Rhdf5lib           1.22.0    2023-04-25 [1] RSPM (R 4.3.0)
 rjson              0.2.21    2022-01-09 [1] CRAN (R 4.3.0)
 rlang              1.1.1     2023-04-28 [1] CRAN (R 4.3.0)
 rmarkdown          2.21      2023-03-26 [1] CRAN (R 4.3.0)
 rstudioapi         0.14      2022-08-22 [1] CRAN (R 4.3.0)
 Rtsne              0.16      2022-04-17 [1] CRAN (R 4.3.0)
 S4Vectors          0.38.1    2023-05-02 [1] RSPM (R 4.3.0)
 scales             1.2.1     2022-08-20 [1] CRAN (R 4.3.0)
 seriation        * 1.4.2     2023-03-08 [1] CRAN (R 4.3.0)
 sessioninfo        1.2.2     2021-12-06 [1] CRAN (R 4.3.0)
 shape              1.4.6     2021-05-19 [1] CRAN (R 4.3.0)
 shiny            * 1.7.4     2022-12-15 [1] CRAN (R 4.3.0)
 stringi            1.7.12    2023-01-11 [1] CRAN (R 4.3.0)
 stringr          * 1.5.0     2022-12-02 [1] CRAN (R 4.3.0)
 survival           3.5-5     2023-03-12 [1] CRAN (R 4.3.0)
 tibble           * 3.2.1     2023-03-20 [1] CRAN (R 4.3.0)
 tidygraph          1.2.3     2023-02-01 [1] CRAN (R 4.3.0)
 tidyr            * 1.3.0     2023-01-24 [1] CRAN (R 4.3.0)
 tidyselect         1.2.0     2022-10-10 [1] CRAN (R 4.3.0)
 tidyverse        * 2.0.0     2023-02-22 [1] CRAN (R 4.3.0)
 timechange         0.2.0     2023-01-11 [1] CRAN (R 4.3.0)
 TSP                1.2-4     2023-04-04 [1] CRAN (R 4.3.0)
 tweenr             2.0.2     2022-09-06 [1] CRAN (R 4.3.0)
 tzdb               0.3.0     2022-03-28 [1] CRAN (R 4.3.0)
 urlchecker         1.0.1     2021-11-30 [1] CRAN (R 4.3.0)
 usethis            2.1.6     2022-05-25 [1] CRAN (R 4.3.0)
 utf8               1.2.3     2023-01-31 [1] CRAN (R 4.3.0)
 vctrs              0.6.2     2023-04-19 [1] CRAN (R 4.3.0)
 vegan              2.6-4     2022-10-11 [1] CRAN (R 4.3.0)
 viridis            0.6.3     2023-05-03 [1] CRAN (R 4.3.0)
 viridisLite        0.4.2     2023-05-02 [1] CRAN (R 4.3.0)
 withr              2.5.0     2022-03-03 [1] CRAN (R 4.3.0)
 xfun               0.39      2023-04-20 [1] CRAN (R 4.3.0)
 xtable             1.8-4     2019-04-21 [1] CRAN (R 4.3.0)
 XVector            0.40.0    2023-04-25 [1] RSPM (R 4.3.0)
 yaml               2.3.7     2023-01-23 [1] CRAN (R 4.3.0)
 zlibbioc           1.46.0    2023-04-25 [1] RSPM (R 4.3.0)

 [1] /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library

‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>